// builderment-classes@0.0.2/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  Area: () => Area,
  Base: () => Base,
  Decoration: () => Decoration,
  DistanceTier: () => DistanceTier,
  DistanceTiers: () => DistanceTiers,
  Factory: () => Factory,
  Item: () => Item,
  ItemCollector: () => ItemCollector,
  OutputTier: () => OutputTier,
  OutputTiers: () => OutputTiers,
  PowerPlant: () => PowerPlant,
  Price: () => Price,
  Seed: () => Seed,
  Space: () => Space,
  SpeedTier: () => SpeedTier,
  SpeedTiers: () => SpeedTiers,
  Storage: () => Storage,
  StorageTier: () => StorageTier,
  StorageTiers: () => StorageTiers,
  Technology: () => Technology,
  Tier: () => Tier,
  Tiers: () => Tiers,
  TransportationDistance: () => TransportationDistance,
  TransportationSpeed: () => TransportationSpeed,
  currencies: () => currencies_exports,
  decorations: () => decorations_exports,
  factories: () => factories_exports,
  itemCollectors: () => item_collectors_exports,
  items: () => items_exports,
  others: () => storages_exports,
  powerPlants: () => power_plants_exports,
  technologies: () => technologies_exports,
  transportations: () => transportations_exports
});
module.exports = __toCommonJS(index_exports);

// src/classes/Area.ts
var Area = class {
  //// Object Properties
  /** The width. */
  width;
  /** The height. */
  height;
  //* Getters
  /**
   * Area.
   * @readonly
   */
  get area() {
    return this.width * this.height;
  }
  constructor(paramsOrArea) {
    this.width = paramsOrArea.width;
    this.height = paramsOrArea.height;
  }
  //// Object Methods
  /**
   * Determine if this {@link Area `Area`} object is equal to another {@link Area `Area`} object.
   * @param area The other {@link Area `Area`} object.
   * @returns `true` if both {@link Area `Area`} objects are equal, `false` otherwise.
   */
  equals(area) {
    if (this === area) return true;
    return this.width === area.width && this.height === area.height;
  }
};

// src/classes/Price.ts
var Price = class _Price {
  //// Static Methods
  /**
   * Add multiple prices together.
   * @param prices Prices to add.
   * @returns Total price.
   */
  static add(...prices) {
    const total = new _Price();
    for (const price of prices) total.add(price instanceof _Price ? price : new _Price(price));
    return total;
  }
  /**
   * Determine if all prices are equal to each other.
   * @param prices The prices to compare.
   * @returns `true` if all prices are equal, `false` otherwise.
   */
  static equalPrices(...prices) {
    if (!prices.length) return false;
    const basePrice = prices[0] instanceof _Price ? prices[0] : new _Price(prices[0]);
    return prices.slice(1).every((price) => basePrice.equals(price instanceof _Price ? price : new _Price(price)));
  }
  //// Object Properties
  /**
   * The amount of gold.
   * @default 0
   */
  gold;
  /**
   * The amount of gems.
   * @default 0
   */
  gems;
  constructor(paramsOrPrice) {
    if (typeof paramsOrPrice === "undefined") {
      this.gold = 0;
      this.gems = 0;
    } else {
      this.gold = typeof paramsOrPrice.gold === "undefined" ? 0 : paramsOrPrice.gold;
      this.gems = typeof paramsOrPrice.gems === "undefined" ? 0 : paramsOrPrice.gems;
    }
  }
  //// Object Methods
  /**
   * Converts this {@link Price `Price`} object to a string.
   */
  toString() {
    return `Price({ gold: ${this.gold}, gems: ${this.gems} })`;
  }
  /**
   * Add two {@link Price `Price`} objects together. This method mutates the original {@link Price `Price`} object.
   * @param price The other {@link Price `Price`} object to add.
   * @returns The original {@link Price `Price`} object.
   */
  add(price) {
    if (!(this.gold === null && price.gold === null)) this.gold += price.gold;
    if (!(this.gems === null && price.gems === null)) this.gems += price.gems;
    return this;
  }
  /**
   * Add two {@link Price `Price`} objects together. This method creates a new {@link Price `Price`} object.
   * @param price The other {@link Price `Price`} object to add.
   * @returns A new {@link Price `Price`} object.
   */
  toAdd(price) {
    return new _Price({
      gold: this.gold === null && price.gold === null ? null : this.gold + price.gold,
      gems: this.gems === null && price.gems === null ? null : this.gems + price.gems
    });
  }
  /**
   * Determine if this {@link Price `Price`} object and another {@link Price `Price`} object are the equal.
   * @param price The other price.
   * @returns `true` if both prices are the equal, `false` otherwise.
   */
  equals(price) {
    if (this === price) return true;
    return this.gold === price.gold && this.gems === price.gems;
  }
  //// Symbols
  /**
   * Called by `Object.prototype.toString`.
   */
  get [Symbol.toStringTag]() {
    return "Price";
  }
  /**
   * Iterator called by the for-of loop.
   */
  *[Symbol.iterator]() {
    yield this.gold;
    yield this.gems;
  }
};

// src/utils/helpers.ts
function toString(any, tabs, limit) {
  if (typeof any === "object") {
    if (tabs.length === limit) return "[object " + any.constructor.name + "]";
    else if (Array.isArray(any)) return arrayToString(any, limit, tabs + "	");
    else return objToString(any, limit, tabs + "	");
  } else if (typeof any === "string") return `"${any}"`;
  else if (typeof any === "bigint") return any + "n";
  return any;
}
function arrayToString(arr, limit = 1, tabs = "	") {
  let string = "[\n";
  for (const value of arr) {
    string += tabs;
    string += toString(value, tabs, limit);
    string += ",\n";
  }
  string = string.slice(0, string.length - 2) + "\n";
  string += (tabs.length > 1 ? tabs.slice(0, tabs.length - 1) : "") + "]";
  if (arr.length === 0) return "[]";
  return string;
}
function objToString(obj, limit = 1, tabs = "	") {
  let string = (obj.constructor.name === "Object" ? "" : obj.constructor.name + " ") + "{\n";
  for (const key in obj) {
    string += tabs + key + ": ";
    string += toString(obj[key], tabs, limit);
    string += ",\n";
  }
  string = string.slice(0, string.length - 2) + "\n";
  string += tabs.slice(0, tabs.length - 1) + "}";
  return string;
}

// src/classes/Base.ts
var Base = class _Base {
  //// Static properties
  //* Public
  /**
   * All the bases that has been created.
   * @readonly
   */
  static bases = {};
  //// Object Properties
  /** The name. */
  name;
  /** The price. */
  price;
  /**
   * URL or relative path to the image.
   * @default "./resources/"
   */
  image;
  constructor(baseOrParams, passByReference = true) {
    this.name = baseOrParams.name;
    if (passByReference && baseOrParams instanceof _Base) this.price = baseOrParams.price;
    else this.price = new Price(baseOrParams.price);
    this.image = baseOrParams.image ?? "./resources/";
    if (typeof _Base.bases[baseOrParams.name] === "undefined") _Base.bases[baseOrParams.name] = [this];
    else _Base.bases[baseOrParams.name].push(this);
  }
  //// Object Methods
  /**
   * Converts this {@link Base `Base`} object into string.
   * @param limit The limit of how many tabs can be used. `limit` must be greater than 0. Default is 2.
   * @returns The string.
   */
  toString(limit = 2) {
    if (limit <= 0) throw new Error("limit must be greater than 0");
    return objToString(this, limit);
  }
  /**
   * Determine if this {@link Base `Base`} object is equal to another {@link Base `Base`} object.
   * @param base The other {@link Base `Base`} object.
   * @returns `true` if both {@link Base `Base`} objects are equal, `false` otherwise.
   */
  equals(base) {
    if (this === base) return true;
    return this.name === base.name && this.price.equals(base.price);
  }
  /**
   * Determine if this {@link Base `Base`} object is strictly equal to another {@link Base `Base`} object.
   * @param base The other {@link Base `Base`} object.
   * @returns `true` if both {@link Base `Base`} objects are strictly equal, `false` otherwise.
   */
  strictlyEquals(base) {
    return this.equals(base) && this.image === base.image;
  }
};

// src/classes/Seed.ts
var import_selfmap = __toESM(require("selfmap"));

// src/objects/buildings/factories.ts
var factories_exports = {};
__export(factories_exports, {
  earthTeleporter: () => earthTeleporter,
  extractor: () => extractor,
  extractorOutputPerMin: () => extractorOutputPerMin,
  forge: () => forge,
  furnace: () => furnace,
  gemTree: () => gemTree,
  gemTreeOutputPerMin: () => gemTreeOutputPerMin,
  industrialFactory: () => industrialFactory,
  machineShop: () => machineShop,
  manufacturer: () => manufacturer,
  uraniumExtractor: () => uraniumExtractor,
  uraniumExtractorOutputPerMin: () => uraniumExtractorOutputPerMin,
  workshop: () => workshop
});

// src/classes/buildings/tiers/Tier.ts
var Tier = class {
  //// Object Properties
  /**
   * The price of that tier.
   * @default 
   * new Price({
   * 	gold: 0,
   * 	gems: 0,
   * })
   */
  price;
  /**
   * The relative path or URL to the image of this tier.
   * @default "./resources/"
   */
  image;
  /**x
   * Construct a {@link Tier `Tier`} object.
   * @param tier A {@link Tier `Tier`} object or tier parameters.
   */
  constructor(paramsOrTier) {
    this.price = paramsOrTier.price instanceof Price ? paramsOrTier.price : typeof paramsOrTier.price === "undefined" ? new Price() : new Price(paramsOrTier.price);
    this.image = paramsOrTier.image ?? "./resources/";
    for (const param in paramsOrTier) {
      if (["price", "image"].indexOf(param) + 1) continue;
      this[param] = paramsOrTier[param];
    }
  }
  //// Object Methods
  /**
   * Converts the tier into string.
   * @param limit The limit of how many tabs can be used. `limit` must be greater than `0`. Default is `2`.
   * @returns The string.
   */
  toString(limit = 2) {
    if (limit <= 0) throw new Error("limit must be greater than 0");
    return objToString(this, limit);
  }
  /**
   * These are similarities between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods.
   * @param tier The other {@link Tier `Tier`} object.
   * @returns `true` if both {@link Tier `Tier`} objects are the equal in the properties that are similar between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods, `false` otherwise.
   */
  similarEquals(tier) {
    return this.price.equals(tier.price);
  }
  /**
   * Determine if this {@link Tier `Tier`} object is equal to another {@link Tier `Tier`} object.
   * @param tier The other {@link Tier `Tier`} object.
   * @returns `true` if both {@link Tier `Tier`} objects are equal, `false` otherwise.
   */
  equals(tier) {
    if (this === tier) return true;
    return this.similarEquals(tier);
  }
  /**
   * Determine if this {@link Tier `Tier`} object is strictly equal to another {@link Tier `Tier`} object.
   * @param tier The other {@link Tier `Tier`} object.
   * @returns `true` if both {@link Tier `Tier`} objects are strictly equal, `false` otherwise.
   */
  strictlyEquals(tier) {
    if (this === tier) return true;
    return this.similarEquals(tier) && this.image === tier.image;
  }
};

// src/classes/buildings/tiers/Tiers.ts
var Tiers = class _Tiers {
  //// Object Properties
  /** All tiers. */
  tiers;
  /**
   * The total price to get to the maximum tier.
   * @readonly
   */
  get maxPrice() {
    return this.getTotalPrice(this.maxTierNum);
  }
  /**
   * Maximum tier number.
   * @readonly
   */
  get maxTierNum() {
    return Math.max(...Object.keys(this.tiers).map((tier) => +tier));
  }
  /**
   * Minimum tier number.
   * @readonly
   */
  get minTierNum() {
    return Math.min(...Object.keys(this.tiers).map((tier) => +tier));
  }
  /**
   * Maximum tier.
   * @readonly
   */
  get maxTier() {
    return this.tiers[this.maxTierNum];
  }
  /**
   * Minimum tier.
   * @readonly
   */
  get minTier() {
    return this.tiers[this.minTierNum];
  }
  constructor(paramsOrTiers, passByReference = true) {
    if (paramsOrTiers instanceof _Tiers) {
      if (passByReference) this.tiers = paramsOrTiers.tiers;
      else {
        this.tiers = {};
        for (const tier in paramsOrTiers.tiers) this.tiers[tier] = new Tier(paramsOrTiers.tiers[tier]);
      }
    } else {
      const minTier = Math.min(...Object.keys(paramsOrTiers).map((tier) => +tier));
      if (Math.trunc(minTier) !== minTier) throw new Error(`All tiers must be integers.`);
      this.tiers = {};
      const maxTier = Math.max(...Object.keys(paramsOrTiers).map((tier) => +tier));
      for (let i = minTier; i <= maxTier; i++) {
        const tier = paramsOrTiers[i];
        if (!tier) throw new Error(`The number ${i} is missing from the tiers. The tiers must include all integers from the smallest value (${this.minTierNum}) to the largest value (${this.maxTierNum}).`);
        if (tier instanceof Tier) this.tiers[i] = passByReference ? tier : new Tier(tier);
        else this.tiers[i] = new Tier(tier);
      }
    }
  }
  //// Object Methods
  /**
   * Converts the tiers into string.
   * @param limit The limit of how many tabs can be used. `limit` must be greater than `0`. Default is `2`.
   * @returns The string.
   */
  toString(limit = 2) {
    if (limit <= 0) throw new Error("limit must be greater than 0");
    return objToString(this, limit);
  }
  /**
   * Check if a tier number is in this {@link Tiers `Tiers`} object's tiers.
   * @param tier Tier number to search for.
   * @param error Whether to throw an error.
   * @throws Error if tier number was not found.
   * @returns Whether the tier number is in this {@link Tiers `Tiers`} object's tiers.
   */
  hasN(tier, error = false) {
    if (!(tier in this.tiers)) {
      if (!error) return false;
      if (tier < this.minTierNum) throw new Error(`"tier" (${tier}) cannot be less than minimum tier (${this.minTierNum}).`);
      if (tier > this.minTierNum) throw new Error(`"tier" (${tier}) cannot be greater than maximum tier (${this.maxTierNum}).`);
      throw new Error(`"tier" (${tier}) cannot be a non-integer number.`);
    }
    return true;
  }
  /**
   * Check if this {@link Tiers `Tiers`} object has a {@link Tier `Tier`} object as one of its tiers.
   * @param tier The {@link Tier `Tier`} object to check.
   * @returns `true` if this {@link Tiers `Tiers`} object has {@link tier `tier`} as one of its tiers, `false` otherwise.
   */
  has(tier) {
    for (const t in this.tiers) if (this.tiers[t].equals(tier)) return true;
    return false;
  }
  /**
   * Check if this {@link Tiers `Tiers`} object strictly has a {@link Tier `Tier`} object as one of its tiers.
   * @param tier The {@link Tier `Tier`} object to check.
   * @returns `true` if this {@link Tiers `Tiers`} object strictly has {@link tier `tier`} as one of its tiers, `false` otherwise.
   */
  strictlyHas(tier) {
    for (const t in this.tiers) if (this.tiers[t].strictlyEquals(tier)) return true;
    return false;
  }
  /**
   * Check if this {@link Tiers `Tiers`} object has a reference to {@link Tier `Tier`} object as one of its tiers.
   * @param tier The {@link Tier `Tier`} object to check.
   * @returns `true` if this {@link Tiers `Tiers`} object has a reference to {@link tier `tier`} as one of its tiers, `false` otherwise.
   */
  hasReference(tier) {
    for (const t in this.tiers) if (this.tiers[t] === tier) return true;
    return false;
  }
  /**
   * Get the total price to get to a tier.
   * @param tier The tier.
   */
  getTotalPrice(tier) {
    this.hasN(tier, true);
    const total = new Price();
    for (let i = this.minTierNum; i <= tier; i++) total.add(this.tiers[i].price);
    return total;
  }
  /**
   * These are similarities between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods.
   * @param tiers The other {@link Tiers `Tiers`} object.
   * @returns `true` if both {@link Tiers `Tiers`} objects are the equal in the properties that are similar between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods, `false` otherwise.
   */
  similarEquals(tiers) {
    if (this === tiers) return true;
    return Object.keys(this.tiers).length === Object.keys(tiers.tiers).length;
  }
  /**
   * Determine if this {@link Tiers `Tiers`} object is equal to another {@link Tiers `Tiers`} object.
   * @param tiers The other {@link Tiers `Tiers`} object.
   * @returns `true` if both tiers are equal, `false` otherwise.
   */
  equals(tiers) {
    if (!this.similarEquals(tiers)) return false;
    for (const tier in this.tiers) if (!this.tiers[tier].equals(tiers.tiers[tier])) return false;
    return true;
  }
  /**
   * Determine if this {@link Tiers `Tiers`} object is strictly equal to another {@link Tiers `Tiers`} object.
   * @param tiers The other {@link Tiers `Tiers`}.
   * @returns `true` if both tiers are strictly equal, `false` otherwise
   */
  strictlyEquals(tiers) {
    if (!this.similarEquals(tiers)) return false;
    for (const tier in this.tiers) if (!this.tiers[tier].strictlyEquals(tiers.tiers[tier])) return false;
    return true;
  }
};

// src/classes/buildings/tiers/output-tiers/OutputTier.ts
var OutputTier = class extends Tier {
  //// Object Properties
  /** The output at this tier. */
  output;
  /**
   * Constructs a new {@link OutputTier `OutputTier`} object.
   * @param paramsOrOutputTier A {@link OutputTier `OutputTier`} object or output tier parameters.
   */
  constructor(paramsOrOutputTier) {
    super(paramsOrOutputTier);
    this.output = paramsOrOutputTier.output;
  }
  //// Object Methods
  /**
   * These are similarities between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods.
   * @param outputTier The other {@link OutputTier `OutputTier`} object.
   * @returns `true` if both {@link OutputTier `OutputTier`} objects are the equal in the properties that are similar between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods, `false` otherwise.
   */
  similarEquals(outputTier) {
    return this.output === outputTier.output;
  }
  /**
   * Determine if this {@link OutputTier `OutputTier`} object is equal to another {@link OutputTier `OutputTier`} object.
   * @param outputTier The other {@link OutputTier `OutputTier`} object.
   * @returns `true` if both {@link OutputTier `OutputTier`} objects are equal, `false` otherwise.
   */
  equals(outputTier) {
    return super.equals(outputTier) && this.similarEquals(outputTier);
  }
  /**
   * Determine if this {@link OutputTier `OutputTier`} object is strictly equal to another {@link OutputTier `OutputTier`} object.
   * @param outputTier The other {@link OutputTier `OutputTier`} object.
   * @returns `true` if both {@link OutputTier `OutputTier`} objects are strictly equal, `false` otherwise.
   */
  strictlyEquals(outputTier) {
    return super.strictlyEquals(outputTier) && this.similarEquals(outputTier);
  }
};

// src/classes/buildings/tiers/output-tiers/OutputTiers.ts
var OutputTiers = class _OutputTiers extends Tiers {
  get maxTier() {
    return this.tiers[this.maxTierNum];
  }
  get minTier() {
    return this.tiers[this.minTierNum];
  }
  constructor(paramsOrOutputTiers, passByReference = true) {
    super(paramsOrOutputTiers, passByReference);
    if (paramsOrOutputTiers instanceof _OutputTiers) {
      if (passByReference) this.tiers = paramsOrOutputTiers.tiers;
      else for (const tier in this.tiers) this.tiers[tier] = new OutputTier(paramsOrOutputTiers.tiers[tier]);
    } else {
      for (const tier in this.tiers) {
        const t = paramsOrOutputTiers[tier];
        if (t instanceof OutputTier) this.tiers[tier] = passByReference ? t : new OutputTier(t);
        else this.tiers[tier] = new OutputTier(t);
      }
    }
  }
  //// Object Methods
  /**
   * These are similarities between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods.
   * @param outputTiers The other {@link OutputTiers `OutputTiers`} object.
   * @returns `true` if both {@link OutputTiers `OutputTiers`} objects are the equal in the things that are similar between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods, `false` otherwise.
   */
  similarEquals(outputTiers) {
    for (const tier in this.tiers) if (this.tiers[tier].output !== outputTiers.tiers[tier].output) return false;
    return true;
  }
  /**
   * Determine if this {@link OutputTiers `OutputTiers`} object is equal to another {@link OutputTiers `OutputTiers`} object.
   * @param outputTiers The other {@link OutputTiers `OutputTiers`} object.
   * @returns `true` if both {@link OutputTiers `OutputTiers`} objects are equal, `false` otherwise.
   */
  equals(outputTiers) {
    return super.equals(outputTiers) && this.similarEquals(outputTiers);
  }
  /**
   * Determine if this {@link OutputTiers `OutputTiers`} object is strictly equal to another {@link OutputTiers `OutputTiers`} object.
   * @param outputTiers The other {@link OutputTiers `OutputTiers`} object.
   * @returns `true` if both {@link OutputTiers `OutputTiers`} objects are strictly equal, `false` otherwise.
   */
  strictlyEquals(outputTiers) {
    return super.strictlyEquals(outputTiers) && this.similarEquals(outputTiers);
  }
};

// src/classes/buildings/tiers/output-tiers/Factory.ts
var Factory = class _Factory extends OutputTiers {
  //// Static properties
  /** 
   * All the factories that has been created.
   * @readonly
   */
  static factories = [];
  //// Object Properties
  /** The name of the factory. */
  name;
  /** The price of the first tier of the factory. */
  price;
  /** The amount of inputs of the factory. */
  inputs;
  /** The description of the factory. */
  description;
  constructor(paramsOrFactory, passByReference = true) {
    super(paramsOrFactory.tiers, passByReference);
    this.name = paramsOrFactory.name;
    this.price = this.tiers[this.minTierNum].price;
    this.inputs = paramsOrFactory.inputs;
    this.description = paramsOrFactory.description;
    if (paramsOrFactory instanceof _Factory) for (const tier in this.tiers) this.tiers[tier].image = paramsOrFactory.tiers[tier].image ?? this.tiers[tier].image + `${this.name} Tier ${tier}.png`;
    else {
      if (paramsOrFactory.tiers instanceof OutputTiers) for (const tier in this.tiers) this.tiers[tier].image = paramsOrFactory.tiers.tiers[tier].image ?? this.tiers[tier].image + `${this.name} Tier ${tier}.png`;
      else for (const tier in this.tiers) this.tiers[tier].image = paramsOrFactory.tiers[tier].image ?? this.tiers[tier].image + `factories/${this.name} Tier ${tier}.png`;
    }
    _Factory.factories.push(this);
  }
  //// Object Methods
  /**
   * These are similarities between the equals and strictlyEquals methods
   * @param factory The other factory
   * @returns true if both factories are the equal in the things that are similar between the equals and strictlyEquals methods, false otherwise
   */
  similarEquals(factory) {
    return this.name === factory.name && this.description === factory.description && this.inputs === factory.inputs;
  }
  /**
   * Determine if this factory and another factory are the equal
   * @param factory The other factory
   * @returns true if both factories are the equal, false otherwise
   */
  equals(factory) {
    return super.equals(factory) && this.similarEquals(factory);
  }
  /**
   * Determine if this factory and another factory are the strictly equal
   * @param factory The other factory
   * @returns true if both factories are the strictly equal, false otherwise
   */
  strictlyEquals(factory) {
    return super.strictlyEquals(factory) && this.similarEquals(factory);
  }
};

// src/objects/buildings/factories.ts
var extractor = new Factory({
  name: "Extractor",
  inputs: 0,
  description: "Place on resource nodes to extract infinite resources.",
  tiers: {
    1: {
      output: 1,
      price: { gold: 10 }
    },
    2: {
      output: 1.5,
      price: { gold: 200 }
    },
    3: {
      output: 2,
      price: { gold: 1e3 }
    },
    4: {
      output: 3,
      price: { gold: 5e3 }
    },
    5: {
      output: 4,
      price: { gold: 2e4 }
    }
  }
});
var extractorOutputPerMin = 7.5;
var uraniumExtractor = new Factory({
  name: "Uranium Extractor",
  inputs: 0,
  description: "Place on resource nodes to extract infinite resources.",
  tiers: {
    1: {
      output: 1,
      price: { gold: 23500 }
    }
  }
});
var uraniumExtractorOutputPerMin = 10;
var workshop = new Factory({
  name: "Workshop",
  inputs: 1,
  description: "Used to craft items automatically, select a recipe after building.",
  tiers: {
    1: {
      output: 1,
      price: { gold: 10 }
    },
    2: {
      output: 1.5,
      price: { gold: 1e3 }
    },
    3: {
      output: 2,
      price: { gold: 4e3 }
    },
    4: {
      output: 3,
      price: { gold: 12e3 }
    }
  }
});
var furnace = new Factory({
  name: "Furnace",
  inputs: 1,
  description: "Used to craft items automatically, select a recipe after building.",
  tiers: {
    1: {
      output: 1,
      price: { gold: 30 }
    },
    2: {
      output: 1.5,
      price: { gold: 200 }
    },
    3: {
      output: 2,
      price: { gold: 800 }
    },
    4: {
      output: 3,
      price: { gold: 3200 }
    }
  }
});
var machineShop = new Factory({
  name: "Machine Shop",
  inputs: 2,
  description: "Used to craft items automatically, select a recipe after building.",
  tiers: {
    1: {
      output: 1,
      price: { gold: 100 }
    },
    2: {
      output: 1.5,
      price: { gold: 500 }
    },
    3: {
      output: 2,
      price: { gold: 2e3 }
    },
    4: {
      output: 3,
      price: { gold: 8e3 }
    }
  }
});
var forge = new Factory({
  name: "Forge",
  inputs: 2,
  description: "Used to craft items automatically, select a recipe after building.",
  tiers: {
    1: {
      output: 1,
      price: { gold: 100 }
    },
    2: {
      output: 1.5,
      price: { gold: 500 }
    },
    3: {
      output: 2,
      price: { gold: 2e3 }
    },
    4: {
      output: 3,
      price: { gold: 1e4 }
    }
  }
});
var industrialFactory = new Factory({
  name: "Industrial Factory",
  inputs: 3,
  description: "Used to craft items automatically, select a recipe after building.",
  tiers: {
    1: {
      output: 1,
      price: { gold: 300 }
    },
    2: {
      output: 1.5,
      price: { gold: 1e3 }
    },
    3: {
      output: 2,
      price: { gold: 4e3 }
    },
    4: {
      output: 3,
      price: { gold: 16e3 }
    }
  }
});
var manufacturer = new Factory({
  name: "Manufacturer",
  inputs: 4,
  description: "Used to craft items automatically, select a recipe after building.",
  tiers: {
    1: {
      output: 1,
      price: { gold: 1e3 }
    },
    2: {
      output: 1.5,
      price: { gold: 4e3 }
    },
    3: {
      output: 2,
      price: { gold: 2e4 }
    },
    4: {
      output: 3,
      price: { gold: 8e4 }
    }
  }
});
var earthTeleporter = new Factory({
  name: "Earth Teleporter",
  inputs: 1,
  description: "Used to craft items automatically, select a recipe after building.",
  tiers: {
    1: {
      output: 1,
      price: { gold: 1e5 }
    }
  }
});
var gemTree = new Factory({
  name: "Gem Tree",
  description: "Money might not grow on trees but Gems do. Gem Apples can be harvested using a Robotic Arm. Send Gem Apples into the Research Lab or Gold Vault to get Gems.",
  inputs: 0,
  tiers: {
    1: {
      output: 1,
      price: { gems: 250 }
    }
  }
});
var gemTreeOutputPerMin = 0.2;

// src/classes/Seed.ts
function hexByteValue(array) {
  let hex = "";
  array.forEach((num) => {
    let h = num.toString(16);
    hex += (h.length !== 2 ? "0" : "") + h;
  });
  return parseInt(hex, 16);
}
var Seed = class _Seed {
  //// Static Properties
  /**
   * Keep track of newly created seeds in the {@link seeds seeds} static variable
   * @default false
   */
  static keepTrackOfSeeds = false;
  /**
   * All the seeds that has been created with a seed.
   * @readonly
   */
  static seeds = new import_selfmap.default([], "seed");
  //// Static Methods
  // TODO: This can be remade to make it more efficient
  /**
   * Get the amount of each raw resource needed for it to NOT be the limited resource
   * @param item The {@link Item `Item`} object to calculate the limited resource based on the base resource of that item
   * @param resources The resources of the seed to calculate the limited resources
   * @returns The amount of each raw resource needed for it to NOT be the limited resource
   */
  static getLimitedDeposits(item, resources) {
    let baseResources = item.getAmountOfBaseResources();
    const limitedAmounts = {};
    for (const key in baseResources) {
      const k = key;
      if (!resources[k]) throw new Error(`Resource (${k}) cannot be 0, null, or undefined`);
      limitedAmounts[k] = resources[k] / baseResources[k];
    }
    const amounts = [];
    for (let i = Object.keys(limitedAmounts).length; i > 0; i--) {
      const min = Math.min(...Object.values(limitedAmounts));
      const name = Object.keys(limitedAmounts).find((key) => limitedAmounts[key] === min);
      delete limitedAmounts[name];
      amounts.push({
        item: Item.getItemsByName(name)[0],
        amount: min
      });
    }
    for (let i = 0; i < amounts.length - 1; i++) {
      amounts[i].amount = amounts[i + 1].amount * baseResources[amounts[i].item.name] - resources[amounts[i].item.name];
    }
    amounts[amounts.length - 1].amount = 0;
    return amounts;
  }
  /**
   * Get the resource that limits making more of the item in the seed
   * @param item The {@link Item `Item`} object to get the resources that limits making more of this item
   * @param resources The resources in the seed
   * @returns The limited resource
   */
  static getLimitedDeposit(item, resources) {
    return _Seed.getLimitedDeposits(item, resources)[0];
  }
  // TODO: Add support for constructing Seed object from .sav file
  /**
   * **This function is incomplete. Do not use.**
   * 
   * Construct a {@link Seed} object from a world save file
   * @param path The path to the world save file
   * @returns A {@link Seed} object
   */
  static async constructFromFile(path) {
    const data = new Uint8Array(await (await fetch(path)).arrayBuffer());
    console.log(data);
    const depositsInfoLocationByte = [22, 26];
    const depositsInfoLocation = [0, 0];
    const seedParams = {
      resources: {
        "Copper Ore": 0,
        "Iron Ore": 0,
        "Uranium Ore": 0,
        "Wood Log": 0,
        Coal: 0,
        Stone: 0,
        Wolframite: 0
      }
    };
    console.log(data.slice(...depositsInfoLocationByte));
    depositsInfoLocation[0] = hexByteValue([...data.slice(...depositsInfoLocationByte)]) + 1;
    depositsInfoLocation[1] = depositsInfoLocation[0] + 71;
    console.log(depositsInfoLocation);
    return new _Seed(seedParams);
  }
  //// Object Properties
  /** The resources in a seed. */
  resources;
  /**
   * The world size of the seed.
   * @default 100
   */
  worldSize;
  /**
   * The resource amount of the seed.
   * @default 100
   */
  resourceAmount;
  /** The seed. */
  seed;
  constructor(seedOrParams, passByReference = true) {
    if (passByReference) this.resources = seedOrParams.resources;
    else this.resources = { ...seedOrParams.resources };
    this.worldSize = seedOrParams.worldSize ?? 100;
    this.resourceAmount = seedOrParams.resourceAmount ?? 100;
    this.seed = seedOrParams.seed;
    if (_Seed.keepTrackOfSeeds && typeof seedOrParams.seed !== "undefined") _Seed.seeds.add(this);
  }
  //// Object Methods
  /**
   * Converts the seed into string
   * @param limit The limit of how many tabs can be used. `limit` must be greater than 0. Default is 2
   * @returns The string
   */
  toString(limit = 2) {
    if (limit <= 0) throw new Error("limit must be greater than 0");
    return objToString(this, limit);
  }
  /**
   * Get the amount of each raw resource needed for it to NOT be the limited resource
   * @param item The {@link Item `Item`} object to calculate the limited resource based on the base resource of that item
   * @returns The amount of each raw resource needed for it to NOT be the limited resource
   */
  getLimitedDeposits(item) {
    return _Seed.getLimitedDeposits(item, this.resources);
  }
  /**
   * Get the resource that limits making more of the item in the seed
   * @param item The {@link Item `Item`} object to get the resources that limits making more of this item
   * @returns The limited resource
   */
  getLimitedDeposit(item) {
    return _Seed.getLimitedDeposit(item, this.resources);
  }
  /**
   * Get the maximum amount to make an item in the seed
   * @param item The {@link Item `Item`} object to get the max of
   * @returns The maximum amount to make the item in the seed
   */
  getMax(item) {
    let i = item.getAmountOfBaseResources();
    const extractorMaxOutput = extractor.maxTier.output * extractorOutputPerMin;
    const uraniumExtractorMaxOutput = uraniumExtractor.maxTier.output * uraniumExtractorOutputPerMin;
    let maximum = Number.MAX_SAFE_INTEGER;
    for (const key in i) maximum = Math.min(maximum, this.resources[key] * (key === "Uranium Ore" ? uraniumExtractorMaxOutput : extractorMaxOutput) / i[key]);
    return maximum;
  }
  /**
   * Determine if this seed and another seed are the equal
   * @param seed The other seed
   * @returns true if both seeds are the equal, false otherwise
   */
  equals(seed) {
    if (this === seed) return true;
    for (const resource in this.resources) if (this.resources[resource] !== seed.resources[resource]) return false;
    return this.resourceAmount === seed.resourceAmount && this.worldSize === seed.worldSize;
  }
  /**
   * Determine if this seed and another seed are the strictly equal
   * @param seed The other seed
   * @returns true if both seeds are the strictly equal, false otherwise
   */
  strictlyEquals(seed) {
    return this.equals(seed) && this.seed === seed.seed;
  }
};

// src/objects/buildings/power-plants.ts
var power_plants_exports = {};
__export(power_plants_exports, {
  coalPowerPlant: () => coalPowerPlant,
  nuclearPowerPlant: () => nuclearPowerPlant
});

// src/classes/buildings/PowerPlant.ts
var PowerPlant = class _PowerPlant extends Base {
  //// Static Properties
  /** 
   * All the power plants that has been created.
   * @readonly
   */
  static powerPlants = [];
  /**
   * The description of all power plants.
   * @readonly
   */
  static get description() {
    return "Increases the speed of nearby factories while consuming the input items. Effect does not stack with other power plants.";
  }
  //// Object Properties
  /** The input needed to activate the power plant. */
  input;
  /** The boost speed that the power plant gives. */
  speed;
  /** The amount of seconds the power plant is active before needing to recharge. */
  duration;
  /** The region of the power plant boost. */
  region;
  constructor(paramsOrPowerPlant, passByReference = true) {
    super(paramsOrPowerPlant, passByReference);
    this.image = paramsOrPowerPlant.image ?? this.image + `power-plants/${this.name}.png`;
    this.input = paramsOrPowerPlant.input;
    this.speed = paramsOrPowerPlant.speed;
    this.duration = paramsOrPowerPlant.duration;
    this.region = passByReference && paramsOrPowerPlant.region instanceof Area ? paramsOrPowerPlant.region : new Area(paramsOrPowerPlant.region);
    _PowerPlant.powerPlants.push(this);
  }
  //// Object Methods
  /**
   * These are similarities between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods.
   * @param powerPlant The other {@link PowerPlant `PowerPlant`} object.
   * @returns `true` if both {@link PowerPlant `PowerPlant`} objects are the equal in the things that are similar between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods, `false` otherwise.
   */
  similarEquals(powerPlant) {
    if (typeof this.input === "undefined") {
      if (typeof powerPlant.input !== "undefined") return false;
    } else if (typeof powerPlant.input === "undefined") return false;
    return this.input.amount === powerPlant.input.amount && this.input.inputPerMin === powerPlant.input.inputPerMin && this.speed === powerPlant.speed && this.duration === powerPlant.duration && this.region.width === powerPlant.region.width && this.region.height === powerPlant.region.height;
  }
  /**
   * Determine if this {@link PowerPlant `PowerPlant`} object is equal to another {@link PowerPlant `PowerPlant`} object.
   * @param powerPlant The other {@link PowerPlant `PowerPlant`} object.
   * @returns `true` if both {@link PowerPlant `PowerPlant`} objects are equal, `false` otherwise.
   */
  equals(powerPlant) {
    return super.equals(powerPlant) && this.similarEquals(powerPlant) && this.input.item.equals(powerPlant.input.item);
  }
  /**
   * Determine if this {@link PowerPlant `PowerPlant`} object is strictly equal to another {@link PowerPlant `PowerPlant`} object.
   * @param powerPlant The other {@link PowerPlant `PowerPlant`} object.
   * @returns `true` if both {@link PowerPlant `PowerPlant`} objects are strictly equal, `false` otherwise.
   */
  strictlyEquals(powerPlant) {
    return super.strictlyEquals(powerPlant) && this.similarEquals(powerPlant) && this.input.item.strictlyEquals(powerPlant.input.item);
  }
};

// src/objects/buildings/power-plants.ts
var coalPowerPlant = new PowerPlant({
  name: "Coal Power Plant",
  price: { gold: 2e3 },
  duration: 15,
  speed: 1.2,
  region: {
    width: 12,
    height: 12
  }
});
var nuclearPowerPlant = new PowerPlant({
  name: "Nuclear Power Plant",
  price: { gold: 5e5 },
  duration: 60,
  speed: 1.4,
  region: {
    width: 21,
    height: 22
  }
});

// src/classes/Item.ts
var Item = class _Item extends Base {
  //// Static Properties
  /**
   * All the items that has been created. 
   * @readonly
   */
  static items = [];
  //// Static Methods
  /**
   * Get Item objects from a name.
   * @param name The item to get.
   * @returns The items if found or `undefined` if not found.
   */
  static getItemsByName(name) {
    const items = [];
    _Item.items.forEach((item) => {
      if (item.name === name) items.push(item);
    });
    return items;
  }
  //// Object Properties
  /** The factory that makes the item. */
  factory;
  /** The image URL of the recipe of the item. */
  recipeImage;
  /** The output/min of the item in tier 1 of the factory. */
  outputPerMin;
  /**
   * The amount of items produces each manufacture.
   * 
   * Example: Copper wire's output amount is `2`.
   * @default 1
   */
  outputAmount;
  /** The resources needed to make the item. */
  resourcesNeeded;
  /** The base resources needed of 1 of the item. */
  baseResources;
  //* Getters
  /** Get the max output/min of the factory to make this item. */
  get maxOutputPerMin() {
    return this.getOutputPerMin(this.factory.maxTierNum);
  }
  /** Get the resources needed to make the item at the max tier. */
  get maxResourcesNeeded() {
    return this.getResourcesNeeded(this.factory.maxTierNum);
  }
  /**
   * Construct an {@link Item `Item`} object.
   * @param paramsOrItem The item object or item parameters.
   */
  constructor(paramsOrItem, passByReference = true) {
    if (paramsOrItem.factory.inputs !== (Array.isArray(paramsOrItem.resourcesNeeded) ? paramsOrItem.resourcesNeeded.length : paramsOrItem.resourcesNeeded.size)) throw new Error("Total amount of resources needed does not equal to the amount of inputs of the factory.");
    super(paramsOrItem);
    this.image = paramsOrItem.image ?? this.image + `items/${this.name}.png`;
    this.recipeImage = paramsOrItem.recipeImage ?? `./resources/recipes/${this.name}.png`;
    this.outputPerMin = paramsOrItem.outputPerMin;
    this.outputAmount = paramsOrItem.outputAmount ?? 1;
    this.resourcesNeeded = /* @__PURE__ */ new Map();
    if (passByReference) {
      this.factory = paramsOrItem.factory;
      if (paramsOrItem instanceof _Item) this.resourcesNeeded = paramsOrItem.resourcesNeeded;
      else for (const input of paramsOrItem.resourcesNeeded) this.resourcesNeeded.set(input.item, { amount: input.amount, inputPerMin: input.amount * this.outputPerMin });
    } else {
      this.factory = new Factory(paramsOrItem.factory, passByReference);
      if (paramsOrItem instanceof _Item) for (const [inputOption, inputPerMin] of paramsOrItem.resourcesNeeded) this.resourcesNeeded.set(new _Item(inputOption, passByReference), { ...inputPerMin });
      else for (const input of paramsOrItem.resourcesNeeded) this.resourcesNeeded.set(input.item, { amount: input.amount, inputPerMin: input.amount * this.outputPerMin });
    }
    this.baseResources = this.getAmountOfBaseResources();
    _Item.items.push(this);
  }
  //// Object Methods
  /**
   * Check if this item needs another item in it's recipe tree.
   * @param item The item to search for.
   * @param scannedItems The scanned items so that they are skipped.
   * @returns `true` if found, `false` otherwise.
   */
  #needsFunction(item, scannedItems) {
    if (scannedItems.includes(this.name)) return false;
    if (this === item) return true;
    scannedItems.push(this.name);
    if (this.resourcesNeeded.size === 0) return false;
    return !this.resourcesNeeded.keys().every((item2) => !item2.#needsFunction(item2, scannedItems));
  }
  /**
   * Check if this item needs another item in it's recipe tree.
   * @param itemOrName The name of the item or an Item object to search for.
   * @returns `true` if found, `false` otherwise.
   */
  needs(itemOrName) {
    if (typeof itemOrName === "string") {
      const items = _Item.getItemsByName(itemOrName);
      if (items.length === 0) return false;
      return items.values().every((item) => this.#needsFunction(item, []));
    } else return this.#needsFunction(itemOrName, []);
  }
  /**
   * Get the amount of base resources used to make the item.
   * @param amount The amount to be added to the total of the item.
   * @param resources The base resources needed to make the item.
   * @returns The amount of base resources needed to make the item.
   */
  #getAmountOfBaseResourcesFunction(amount, resources = {}) {
    if (this.resourcesNeeded.size) {
      for (const [item, inputPerMin] of this.resourcesNeeded) {
        amount /= this.outputAmount;
        for (let i = 0; i < inputPerMin.amount; i++) {
          resources = item.#getAmountOfBaseResourcesFunction(amount, resources);
        }
      }
    } else {
      if (typeof resources[this.name] === "undefined" || isNaN(resources[this.name])) resources[this.name] = amount;
      else resources[this.name] += amount;
    }
    return resources;
  }
  /**
   * Get the amount of base resources used to make the item.
   * @param amount The amount of that resource. Default is 1.
   */
  getAmountOfBaseResources(amount = 1) {
    const resources = this.#getAmountOfBaseResourcesFunction(amount);
    for (const resource in resources) resources[resource] = Math.round(resources[resource] * 1e3) / 1e3;
    return resources;
  }
  /**
   * Get the amount of resources needed to make the item.
   * @param amount The amount to be added to the total of the item.
   * @param resources The amount of resources needed to make the item.
   * @returns The amount of resources needed to make the item.
   */
  #getAmountOfResourcesFunction(amount, resources = {}) {
    if (isNaN(resources[this.name])) resources[this.name] = 0;
    resources[this.name] += amount;
    for (const [item, inputPerMin] of this.resourcesNeeded) {
      amount /= this.outputAmount;
      for (let i = 0; i < inputPerMin.amount; i++) {
        resources = item.#getAmountOfResourcesFunction(amount, resources);
      }
    }
    return resources;
  }
  /**
   * Get the amount of resources used to make the item.
   * @param amount The amount of that resource. Default is 1.
   */
  getAmountOfResources(amount = 1) {
    const resources = this.#getAmountOfResourcesFunction(amount);
    for (const resource in resources) resources[resource] = Math.round(resources[resource] * 1e3) / 1e3;
    return resources;
  }
  /**
   * Get the output/min based on the tier of the factory.
   * @param tier The tier to get the output of. Default is 1.
   */
  getOutputPerMin(tier = 1) {
    this.factory.hasN(tier, true);
    return this.outputPerMin * this.factory.tiers[tier].output;
  }
  /**
   * Get the resources needed to make the item at a tier
   * @param tier The tier to get the resources needed. Default is 1
   */
  getResourcesNeeded(tier = 1) {
    this.factory.hasN(tier, true);
    const resources = /* @__PURE__ */ new Map();
    for (const [item, inputPerMin] of this.resourcesNeeded) resources.set(item, {
      amount: inputPerMin.amount,
      inputPerMin: this.resourcesNeeded.get(item).inputPerMin * this.factory.tiers[tier].output
    });
    return resources;
  }
  /**
   * Get the maximum amount you can get of this item in a seed.
   * 
   * **Note:** This method only works with the normal in-game items.
   * @param resourcesOrSeed The resources in the world or a {@link Seed `Seed`} object.
   * @returns the maximum amount you can get of this item.
   */
  getMaxResourceAmountInSeed(resourcesOrSeed) {
    const resources = resourcesOrSeed instanceof Seed ? resourcesOrSeed.resources : resourcesOrSeed;
    const extractorMaxOutput = extractor.maxTier.output * extractorOutputPerMin;
    const uraniumExtractorMaxOutput = uraniumExtractor.maxTier.output * uraniumExtractorOutputPerMin;
    let lowest = Number.MAX_SAFE_INTEGER;
    for (const name in this.baseResources) lowest = Math.min(lowest, resources[name] * (name === "Uranium Ore" ? uraniumExtractorMaxOutput : extractorMaxOutput) / this.baseResources[name]);
    return lowest;
  }
  getMaxResourceAmountInSeedWithPowerPlants(resourcesOrSeed, resourceAmount = 100) {
    let coalPowerPlantExtractorReach;
    let nuclearPowerPlantExtractorReach;
    let uraniumOreDepositsInPatch;
    switch (resourcesOrSeed instanceof Seed ? resourcesOrSeed.resourceAmount : resourceAmount) {
      case 50:
        coalPowerPlantExtractorReach = NaN;
        nuclearPowerPlantExtractorReach = NaN;
        uraniumOreDepositsInPatch = NaN;
        break;
      case 75:
        coalPowerPlantExtractorReach = NaN;
        nuclearPowerPlantExtractorReach = NaN;
        uraniumOreDepositsInPatch = NaN;
        break;
      case 100:
        coalPowerPlantExtractorReach = 22;
        nuclearPowerPlantExtractorReach = 36;
        uraniumOreDepositsInPatch = 6.9;
        break;
      case 150:
        coalPowerPlantExtractorReach = NaN;
        nuclearPowerPlantExtractorReach = NaN;
        uraniumOreDepositsInPatch = 6.9;
        break;
      case 200:
        coalPowerPlantExtractorReach = 22;
        nuclearPowerPlantExtractorReach = 55;
        uraniumOreDepositsInPatch = 6.9;
        break;
    }
    const resources = resourcesOrSeed instanceof Seed ? resourcesOrSeed.resources : resourcesOrSeed;
    const r = { ...resources };
    const extractorMaxOutput = extractor.maxTier.output * extractorOutputPerMin;
    const uraniumExtractorMaxOutput = uraniumExtractor.maxTier.output * uraniumExtractorOutputPerMin;
    const nuclearFuelCellBaseResources = nuclearPowerPlant.input.item.baseResources;
    r.Coal -= r["Uranium Ore"] / uraniumOreDepositsInPatch;
    r["Uranium Ore"] *= coalPowerPlant.speed;
    r.Coal = r.Coal * coalPowerPlant.speed - r.Coal / coalPowerPlantExtractorReach * (coalPowerPlant.input.inputPerMin / extractorMaxOutput);
    while (r["Uranium Ore"] - nuclearFuelCellBaseResources["Uranium Ore"] / uraniumExtractorMaxOutput > 0) {
      for (const resource in nuclearFuelCellBaseResources) r[resource] -= nuclearFuelCellBaseResources[resource] / (resource === "Uranium Ore" ? uraniumExtractorMaxOutput : extractorMaxOutput);
      const limited = Seed.getLimitedDeposit(this, r);
      if (limited.item.name === "Coal" || limited.item.name === "Uranium Ore") break;
      r[limited.item.name] += nuclearPowerPlantExtractorReach * nuclearPowerPlant.speed - nuclearPowerPlantExtractorReach;
    }
    while (true) {
      const limited = Seed.getLimitedDeposit(this, { ...r, Coal: r.Coal - coalPowerPlant.input.inputPerMin / extractorMaxOutput });
      if (limited.item.name === "Coal" || limited.item.name === "Uranium Ore") break;
      r.Coal -= coalPowerPlant.input.inputPerMin / extractorMaxOutput;
      r[limited.item.name] += coalPowerPlantExtractorReach * coalPowerPlant.speed - coalPowerPlantExtractorReach;
    }
    return this.getMaxResourceAmountInSeed(r);
  }
  /**
   * These are similarities between the equals and strictlyEquals methods.
   * @param item The other item.
   * @returns `true` if both items are the equal in the things that are similar between the equals and strictlyEquals methods, `false` otherwise.
   */
  similarEquals(item) {
    return this.resourcesNeeded.size === item.resourcesNeeded.size && this.outputAmount === item.outputAmount && this.outputPerMin === item.outputPerMin;
  }
  /**
   * Determine if this {@link Item `Item`} object and another {@link Item `Item`} object are equal.
   * @param item The other {@link Item `Item`} object.
   * @returns `true` if both items are the equal, `false` otherwise.
   */
  equals(item) {
    if (this === item) return true;
    for (const [thisResourcesNeededItem, inputPerMin] of this.resourcesNeeded) if (inputPerMin.amount !== item.resourcesNeeded.get(thisResourcesNeededItem).amount || // Check amount
    inputPerMin.inputPerMin !== item.resourcesNeeded.get(thisResourcesNeededItem).inputPerMin || // Check inputPerMin
    !item.resourcesNeeded.keys().some((itemResourceNeededItem) => thisResourcesNeededItem.equals(itemResourceNeededItem))) return false;
    return super.equals(item) && this.similarEquals(item) && this.factory.equals(item.factory);
  }
  /**
   * Determine if this {@link Item `Item`} object and another {@link Item `Item`} object are strictly equal.
   * @param item The other {@link Item `Item`} object.
   * @returns `true` if both items are the strictly equal, `false` otherwise.
   */
  strictlyEquals(item) {
    if (this === item) return true;
    for (const [thisResourcesNeededItem, inputPerMin] of this.resourcesNeeded) if (inputPerMin.amount !== item.resourcesNeeded.get(thisResourcesNeededItem).amount || // Check amount
    inputPerMin.inputPerMin !== item.resourcesNeeded.get(thisResourcesNeededItem).inputPerMin || // Check inputPerMin
    !item.resourcesNeeded.keys().some((itemResourceNeededItem) => thisResourcesNeededItem.strictlyEquals(itemResourceNeededItem))) return false;
    return super.strictlyEquals(item) && this.similarEquals(item) && this.factory.strictlyEquals(item.factory) && this.recipeImage === item.recipeImage;
  }
};

// src/classes/buildings/ItemCollector.ts
var ItemCollector = class _ItemCollector extends Base {
  //// Static Properties
  /**
   * All the item collectors that has been created.
   * @readonly
   */
  static itemCollectors = [];
  //// Object Properties
  /** The description of the build. */
  description;
  /** The limit how many of the build can be build. */
  limit;
  /** The amount of inputs of the build. */
  inputs;
  constructor(itemCollectorOrOptions) {
    super(itemCollectorOrOptions);
    this.image = itemCollectorOrOptions.image ?? this.image + `item-collectors/${this.name}.png`;
    this.description = itemCollectorOrOptions.description;
    this.limit = itemCollectorOrOptions.limit;
    this.inputs = itemCollectorOrOptions.inputs;
    _ItemCollector.itemCollectors.push(this);
  }
  //// Object Methods
  /**
   * These are similarities between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods.
   * @param itemCollector The other {@link ItemCollector `ItemCollector`} object.
   * @returns `true` if both {@link ItemCollector `ItemCollector`} objects are the equal in the things that are similar between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods, `false` otherwise.
   */
  similarEquals(itemCollector) {
    return this.description === itemCollector.description && this.limit === itemCollector.limit && this.inputs === itemCollector.inputs;
  }
  /**
   * Determine if this {@link ItemCollector `ItemCollector`} object is equal to another {@link ItemCollector `ItemCollector`} object.
   * @param itemCollector The other {@link ItemCollector `ItemCollector`} object.
   * @returns `true` if both {@link ItemCollector `ItemCollector`} objects are equal, `false` otherwise.
   */
  equals(itemCollector) {
    return super.equals(itemCollector) && this.similarEquals(itemCollector);
  }
  /**
   * Determine if this {@link ItemCollector `ItemCollector`} object is strictly equal to another {@link ItemCollector `ItemCollector`} object.
   * @param itemCollector The other {@link ItemCollector `ItemCollector`} object.
   * @returns `true` if both {@link ItemCollector `ItemCollector`} objects are strictly equal, `false` otherwise.
   */
  strictlyEquals(itemCollector) {
    return super.strictlyEquals(itemCollector) && this.similarEquals(itemCollector);
  }
};

// src/classes/buildings/tiers/storage-tiers/StorageTier.ts
var StorageTier = class extends Tier {
  //// Object Properties
  /** The output at this tier. */
  storage;
  /**
   * Constructs a new {@link StorageTier `StorageTier`} object.
   * @param storageTier A {@link StorageTier `StorageTier`} object or storage tier parameters.
   */
  constructor(paramsOrStorageTier) {
    super(paramsOrStorageTier);
    this.storage = paramsOrStorageTier.storage;
  }
  //// Object Methods
  /**
   * These are similarities between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods.
   * @param distanceTier The other {@link StorageTier `StorageTier`} object.
   * @returns `true` if both {@link StorageTier `StorageTier`} objects are the equal in the properties that are similar between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods, `false` otherwise.
   */
  similarEquals(storageTier) {
    return this.storage === storageTier.storage;
  }
  /**
   * Determine if this {@link StorageTier `StorageTier`} object is equal to another {@link StorageTier `StorageTier`} object.
   * @param storageTier The other {@link StorageTier `StorageTier`} object.
   * @returns `true` if both {@link StorageTier `StorageTier`} objects are equal, `false` otherwise.
   */
  equals(storageTier) {
    return super.equals(storageTier) && this.similarEquals(storageTier);
  }
  /**
   * Determine if this {@link StorageTier `StorageTier`} object is strictly equal to another {@link StorageTier `StorageTier`} object.
   * @param storageTier The other {@link StorageTier `StorageTier`} object.
   * @returns `true` if both {@link StorageTier `StorageTier`} objects are strictly equal, `false` otherwise.
   */
  strictlyEquals(storageTier) {
    return super.strictlyEquals(storageTier) && this.similarEquals(storageTier);
  }
};

// src/classes/buildings/tiers/storage-tiers/StorageTiers.ts
var StorageTiers = class _StorageTiers extends Tiers {
  get maxTier() {
    return this.tiers[this.maxTierNum];
  }
  get minTier() {
    return this.tiers[this.minTierNum];
  }
  constructor(paramsOrStorageTiers, passByReference = true) {
    super(paramsOrStorageTiers, passByReference);
    if (paramsOrStorageTiers instanceof _StorageTiers) {
      if (passByReference) this.tiers = paramsOrStorageTiers.tiers;
      else for (const tier in this.tiers) this.tiers[tier] = new StorageTier(paramsOrStorageTiers.tiers[tier]);
    } else {
      for (const tier in this.tiers) {
        const t = paramsOrStorageTiers[tier];
        if (t instanceof StorageTier) this.tiers[tier] = passByReference ? t : new StorageTier(t);
        else this.tiers[tier] = new StorageTier(t);
      }
    }
  }
  //// Object Methods
  /**
   * These are similarities between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods.
   * @param storageTiers The other {@link StorageTiers `StorageTiers`} object.
   * @returns `true` if both {@link StorageTiers `StorageTiers`} objects are the equal in the things that are similar between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods, `false` otherwise.
   */
  similarEquals(storageTiers) {
    for (const tier in this.tiers) if (this.tiers[tier].storage !== storageTiers.tiers[tier].storage) return false;
    return true;
  }
  /**
   * Determine if this {@link StorageTiers `StorageTiers`} object is equal to another {@link StorageTiers `StorageTiers`} object.
   * @param storageTiers The other {@link StorageTiers `StorageTiers`} object.
   * @returns `true` if both {@link StorageTiers `StorageTiers`} objects are equal, `false` otherwise.
   */
  equals(storageTiers) {
    return super.equals(storageTiers) && this.similarEquals(storageTiers);
  }
  /**
   * Determine if this {@link StorageTiers `StorageTiers`} object is strictly equal to another {@link StorageTiers `StorageTiers`} object.
   * @param storageTiers The other {@link StorageTiers `StorageTiers`} object.
   * @returns `true` if both {@link StorageTiers `StorageTiers`} objects are strictly equal, `false` otherwise.
   */
  strictlyEquals(storageTiers) {
    return super.strictlyEquals(storageTiers) && this.similarEquals(storageTiers);
  }
};

// src/classes/buildings/tiers/storage-tiers/Storage.ts
var Storage = class _Storage extends StorageTiers {
  //// Static properties
  /**
   * All the transportations that has been created.
   * @readonly
   */
  static storages = [];
  //// Object Properties
  /** The name of the storage. */
  name;
  /**
   * The price of the storage.
   * @default
   * ```javascript
   * new Price({
   * 	gold: 0,
   * 	gems: 0
   * })
   * ```
   */
  price;
  /** The description of the storage. */
  description;
  constructor(paramsOrStorage, passByReference = true) {
    super(paramsOrStorage.tiers, passByReference);
    this.name = paramsOrStorage.name;
    this.price = this.tiers[this.minTierNum].price;
    this.description = paramsOrStorage.description;
    if (paramsOrStorage instanceof _Storage) for (const tier in this.tiers) this.tiers[tier].image = paramsOrStorage.tiers[tier].image ?? this.tiers[tier].image + `storages/${this.name} Tier ${tier}.png`;
    else {
      if (paramsOrStorage.tiers instanceof StorageTiers) for (const tier in this.tiers) this.tiers[tier].image = paramsOrStorage.tiers.tiers[tier].image ?? this.tiers[tier].image + `storages/${this.name} Tier ${tier}.png`;
      else for (const tier in this.tiers) this.tiers[tier].image = paramsOrStorage.tiers[tier].image ?? this.tiers[tier].image + `storages/${this.name} Tier ${tier}.png`;
    }
    _Storage.storages.push(this);
  }
  //// Object Methods
  /**
   * These are similarities between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods.
   * @param storage The other {@link Storage `Storage`} object.
   * @returns `true` if both {@link Storage `Storage`} objects are the equal in the things that are similar between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods, `false` otherwise.
   */
  similarEquals(storage3) {
    return this.name === storage3.name && this.price.equals(storage3.price) && this.description === storage3.description;
  }
  /**
   * Determine if this {@link Storage `Storage`} object is equal to another {@link Storage `Storage`} object.
   * @param storage The other {@link Storage `Storage`} object.
   * @returns `true` if both storages are the equal, `false` otherwise.
   */
  equals(storage3) {
    return super.equals(storage3) && this.similarEquals(storage3);
  }
  /**
   * Determine if this {@link Storage `Storage`} object is strictly equal to {@link Storage `Storage`} object.
   * @param storage The other {@link Storage `Storage`} object.
   * @returns `true` if both storages are strictly equal, `false` otherwise.
   */
  strictlyEquals(storage3) {
    return super.strictlyEquals(storage3) && this.similarEquals(storage3);
  }
};

// src/classes/buildings/tiers/distance-tiers/DistanceTier.ts
var DistanceTier = class extends Tier {
  //// Object Properties
  /** The output at this tier. */
  distance;
  /**
   * Constructs a new {@link DistanceTier `DistanceTier`} object.
   * @param paramsOrDistanceTier A {@link DistanceTier `DistanceTier`} object or distance tier parameters.
   */
  constructor(paramsOrDistanceTier) {
    super(paramsOrDistanceTier);
    this.distance = paramsOrDistanceTier.distance;
  }
  //// Object Methods
  /**
   * These are similarities between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods.
   * @param distanceTier The other {@link DistanceTier `DistanceTier`} object.
   * @returns `true` if both {@link DistanceTier `DistanceTier`} objects are the equal in the properties that are similar between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods, `false` otherwise.
   */
  similarEquals(distanceTier) {
    return this.distance === distanceTier.distance;
  }
  /**
   * Determine if this {@link DistanceTier `DistanceTier`} object is equal to another {@link DistanceTier `DistanceTier`} object.
   * @param distanceTier The other {@link DistanceTier `DistanceTier`} object.
   * @returns `true` if both {@link DistanceTier `DistanceTier`} objects are equal, `false` otherwise.
   */
  equals(distanceTier) {
    return super.equals(distanceTier) && this.similarEquals(distanceTier);
  }
  /**
   * Determine if this {@link DistanceTier `DistanceTier`} object is strictly equal to another {@link DistanceTier `DistanceTier`} object.
   * @param distanceTier The other {@link DistanceTier `DistanceTier`} object.
   * @returns `true` if both {@link DistanceTier `DistanceTier`} objects are strictly equal, `false` otherwise.
   */
  strictlyEquals(distanceTier) {
    return super.strictlyEquals(distanceTier) && this.similarEquals(distanceTier);
  }
};

// src/classes/buildings/tiers/distance-tiers/DistanceTiers.ts
var DistanceTiers = class _DistanceTiers extends Tiers {
  constructor(paramsOrDistanceTiers, passByReference = true) {
    super(paramsOrDistanceTiers, passByReference);
    if (paramsOrDistanceTiers instanceof _DistanceTiers) {
      if (passByReference) this.tiers = paramsOrDistanceTiers.tiers;
      else for (const tier in this.tiers) this.tiers[tier] = new DistanceTier(paramsOrDistanceTiers.tiers[tier]);
    } else {
      for (const tier in this.tiers) {
        const t = paramsOrDistanceTiers[tier];
        if (t instanceof DistanceTier) this.tiers[tier] = passByReference ? t : new DistanceTier(t);
        else this.tiers[tier] = new DistanceTier(t);
      }
    }
  }
  //// Object Methods
  /**
   * These are similarities between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods.
   * @param distanceTier The other {@link DistanceTiers `DistanceTiers`} object.
   * @returns `true` if both {@link DistanceTiers `DistanceTiers`} objects are the equal in the things that are similar between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods, `false` otherwise.
   */
  similarEquals(distanceTier) {
    for (const tier in this.tiers) if (this.tiers[tier].distance !== distanceTier.tiers[tier].distance) return false;
    return true;
  }
  /**
   * Determine if this {@link DistanceTiers `DistanceTiers`} object is equal to another {@link DistanceTiers `DistanceTiers`} object.
   * @param distanceTier The other {@link DistanceTiers `DistanceTiers`} object.
   * @returns `true` if both {@link DistanceTiers `DistanceTiers`} objects are equal, `false` otherwise.
   */
  equals(distanceTier) {
    return super.equals(distanceTier) && this.similarEquals(distanceTier);
  }
  /**
   * Determine if this {@link DistanceTiers `DistanceTiers`} object is strictly equal to another {@link DistanceTiers `DistanceTiers`} object.
   * @param distanceTier The other {@link DistanceTiers `DistanceTiers`} object.
   * @returns `true` if both {@link DistanceTiers `DistanceTiers`} objects are strictly equal, `false` otherwise.
   */
  strictlyEquals(distanceTier) {
    return super.strictlyEquals(distanceTier) && this.similarEquals(distanceTier);
  }
};

// src/classes/buildings/tiers/distance-tiers/TransportationDistance.ts
var TransportationDistance = class _TransportationDistance extends DistanceTiers {
  //// Static properties
  /**
   * All the transportation distances that has been created.
   * @readonly
   */
  static transportationDistances = [];
  //// Object Properties
  /** The name of the transportation distance. */
  name;
  /** The price of the first tier of the transportation distance. */
  price;
  /** The description of the transportation distance. */
  description;
  constructor(optionsOrTransportationDistance, passByReference = true) {
    super(optionsOrTransportationDistance.tiers, passByReference);
    this.name = optionsOrTransportationDistance.name;
    this.price = this.tiers[this.minTierNum].price;
    this.description = optionsOrTransportationDistance.description;
    if (optionsOrTransportationDistance instanceof _TransportationDistance) for (const tier in this.tiers) this.tiers[tier].image = optionsOrTransportationDistance.tiers[tier].image ?? this.tiers[tier].image + `transportations/${this.name}.png`;
    else {
      if (optionsOrTransportationDistance.tiers instanceof DistanceTiers) for (const tier in this.tiers) this.tiers[tier].image = optionsOrTransportationDistance.tiers.tiers[tier].image ?? this.tiers[tier].image + `transportations/${this.name}.png`;
      else for (const tier in this.tiers) this.tiers[tier].image = optionsOrTransportationDistance.tiers[tier].image ?? this.tiers[tier].image + `transportations/${this.name}.png`;
    }
    _TransportationDistance.transportationDistances.push(this);
  }
  //// Object Methods
  /**
   * These are similarities between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods.
   * @param transportationDistance The other {@link TransportationDistance `TransportationDistance`} object.
   * @returns `true` if both {@link TransportationDistance `TransportationDistance`} objects are the equal in the things that are similar between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods, `false` otherwise.
   */
  similarEquals(transportationDistance) {
    return this.name === transportationDistance.name && this.price.equals(transportationDistance.price) && this.description === transportationDistance.description;
  }
  /**
   * Determine if this {@link TransportationDistance `TransportationDistance`} object is equal to another {@link TransportationDistance `TransportationDistance`} object.
   * @param transportationDistance The other {@link TransportationDistance `TransportationDistance`} object.
   * @returns `true` if both storages are the equal, `false` otherwise.
   */
  equals(transportationDistance) {
    return super.equals(transportationDistance) && this.similarEquals(transportationDistance);
  }
  /**
   * Determine if this {@link TransportationDistance `TransportationDistance`} object is strictly equal to another {@link TransportationDistance `TransportationDistance`} object.
   * @param transportationDistance The other {@link TransportationDistance `TransportationDistance`} object.
   * @returns `true` if both storages are the strictly equal, `false` otherwise.
   */
  strictlyEquals(transportationDistance) {
    return super.strictlyEquals(transportationDistance) && this.similarEquals(transportationDistance);
  }
};

// src/classes/Technology.ts
var Technology = class _Technology extends Base {
  //// Static Properties
  /**
   * All the technologies that has been created
   * @readonly
   */
  static technologies = [];
  /**
   * The image path of the Tech-tree.
   * @readonly
   */
  static get treeImage() {
    return "./resources/technologies/Tech Tree.png";
  }
  /**
   * Total price of all technologies
   * @readonly
   */
  static get totalPrice() {
    let total = { gold: 0, gems: 0 };
    for (const technology of _Technology.technologies) for (const currency in total) total[currency] += technology.price[currency] ?? 0;
    return total;
  }
  //// Static Methods
  /**
   * Get a {@link Technology `Technology`} object.
   * @param name A technology name.
   * @returns An array of {@link Technology `Technology`} objects if found, else `undefined`.
   */
  static getTechnologiesByName(name) {
    const technologies = [];
    _Technology.technologies.forEach((technology) => {
      if (technology.name === name) technologies.push(technology);
    });
    return technologies;
  }
  //// Object Properties
  /** The items needed to unlock this technology */
  resourcesNeeded;
  /** The items, factories, or builds that this technology unlocks. */
  unlocks;
  /**
   * The technologies that this technology unlocks.
   * @default []
  */
  unlocksTechnologies;
  /**
   * The technology needed to unlock this technology.
   * @default null
   */
  technologyNeeded;
  /**
   * Get the total price to get to this technology.
   * @readonly
   */
  get totalPrice() {
    if (this.technologyNeeded === null) return new Price();
    return Price.add(this.price, this.technologyNeeded.totalPrice);
  }
  constructor(technologyOrOptions, passByReference = true) {
    super(technologyOrOptions);
    const firstUnlock = technologyOrOptions.unlocks[0];
    this.image = technologyOrOptions.image ?? this.image + `technologies/${this.name}` + // is BuildTierOptions
    (Object.hasOwn(firstUnlock, "build") ? (
      // is Transportation and underground belt upgrade
      firstUnlock.build instanceof TransportationDistance && firstUnlock.build.name === "Underground Belt Upgrade" ? ` ${firstUnlock.tier}` : ` Tier ${firstUnlock.tier}`
    ) : "") + ".png";
    if (passByReference) {
      this.technologyNeeded = typeof technologyOrOptions.technologyNeeded === "undefined" ? null : technologyOrOptions.technologyNeeded;
      this.resourcesNeeded = technologyOrOptions.resourcesNeeded;
      this.unlocks = technologyOrOptions.unlocks;
      this.unlocksTechnologies = technologyOrOptions.unlocksTechnologies ?? [];
    } else {
      this.technologyNeeded = typeof technologyOrOptions.technologyNeeded === "undefined" || technologyOrOptions.technologyNeeded === null ? null : new _Technology(technologyOrOptions.technologyNeeded);
      this.resourcesNeeded = [];
      for (const resourceNeeded of technologyOrOptions.resourcesNeeded) this.resourcesNeeded.push({
        amount: resourceNeeded.amount,
        item: new Item(resourceNeeded.item)
      });
      this.unlocks = [];
      for (const unlock of technologyOrOptions.unlocks) this.unlocks.push(unlock instanceof Base ? new Base(unlock) : unlock instanceof ItemCollector ? new ItemCollector(unlock) : {
        tier: unlock.tier,
        build: unlock.build instanceof Factory ? new Factory(unlock.build) : unlock.build instanceof TransportationDistance ? new TransportationDistance(unlock.build) : unlock.build instanceof Storage ? new Storage(unlock.build) : unlock.build instanceof OutputTiers ? new OutputTiers(unlock.build) : unlock.build instanceof StorageTiers ? new StorageTiers(unlock.build) : unlock.build instanceof DistanceTiers ? new DistanceTiers(unlock.build) : new Tiers(unlock.build)
      });
      this.unlocksTechnologies = [];
      if (typeof technologyOrOptions.unlocksTechnologies !== "undefined") for (const unlocksTechnology of technologyOrOptions.unlocksTechnologies) this.unlocksTechnologies.push(unlocksTechnology);
    }
    _Technology.technologies.push(this);
  }
  //// Object Methods
  /**
   * Check if a {@link Technology `Technology`} object is needed to unlock this {@link Technology `Technology`} object.
   * @param technology The {@link Technology `Technology`} object to check if it is needed to unlock this {@link Technology `Technology`} object.
   * @returns `true` if {@link technology `technology`} is needed, `false` otherwise.
   */
  needsFunction(technology) {
    if (this === technology) return true;
    if (typeof this.technologyNeeded === "undefined" || this.technologyNeeded === null) return false;
    return this.technologyNeeded.needsFunction(technology);
  }
  needs(technologyOrName) {
    if (typeof technologyOrName === "string") {
      const technologies = _Technology.getTechnologiesByName(technologyOrName);
      if (technologies.length === 0) return false;
      return technologies.values().every((technology) => this.needsFunction(technology));
    } else return this.needsFunction(technologyOrName);
  }
  /**
   * Determine if this technology and another technology are the equal
   * @param technology The other technology
   * @returns true if both technologies are the equal, false otherwise
   */
  equals(technology) {
    return super.equals(technology) && this.resourcesNeeded.length === technology.resourcesNeeded.length && this.resourcesNeeded.every((input) => technology.resourcesNeeded.some((otherInput) => input.amount === otherInput.amount && input.item.equals(otherInput.item))) && this.unlocks.every((unlock) => technology.unlocks.some((otherUnlock) => {
      if (typeof otherUnlock === "undefined") return false;
      if (unlock instanceof Base) {
        if (!(otherUnlock instanceof Base)) return false;
        if (unlock instanceof ItemCollector) return otherUnlock instanceof ItemCollector && unlock.equals(otherUnlock);
        else return !(otherUnlock instanceof ItemCollector) && unlock.equals(otherUnlock);
      } else {
        if (otherUnlock instanceof Base) return false;
        if (unlock.tier !== otherUnlock.tier) return false;
        if (Object.getPrototypeOf(unlock.build).constructor.name === Object.getPrototypeOf(otherUnlock.build).constructor.name) return unlock.build.equals(otherUnlock.build);
        else return false;
      }
    })) && this.unlocks.length === technology.unlocks.length && (typeof this.unlocksTechnologies === "undefined" ? typeof technology.unlocksTechnologies === "undefined" : typeof technology.unlocksTechnologies === "undefined" ? false : this.unlocksTechnologies.every((tech) => technology.unlocksTechnologies.some((otherTech) => tech.equals(otherTech))));
  }
  /**
   * Determine if this technology and another technology are the strictly equal
   * @param technology The other technology
   * @returns true if both technologies are the strictly equal, false otherwise
   */
  strictlyEquals(technology) {
    return super.strictlyEquals(technology) && this.resourcesNeeded.length === technology.resourcesNeeded.length && this.resourcesNeeded.every((input) => technology.resourcesNeeded.some((otherInput) => input.amount === otherInput.amount && input.item.strictlyEquals(otherInput.item))) && this.unlocks.every((unlock) => technology.unlocks.some((otherUnlock) => {
      if (typeof otherUnlock === "undefined") return false;
      if (unlock instanceof Base) {
        if (!(otherUnlock instanceof Base)) return false;
        if (unlock instanceof ItemCollector) return otherUnlock instanceof ItemCollector && unlock.strictlyEquals(otherUnlock);
        else return !(otherUnlock instanceof ItemCollector) && unlock.strictlyEquals(otherUnlock);
      } else {
        if (otherUnlock instanceof Base) return false;
        if (unlock.tier !== otherUnlock.tier) return false;
        if (Object.getPrototypeOf(unlock.build).constructor.name === Object.getPrototypeOf(otherUnlock.build).constructor.name) return unlock.build.equals(otherUnlock.build);
        else return false;
      }
    })) && this.unlocks.length === technology.unlocks.length && (typeof this.unlocksTechnologies === "undefined" ? typeof technology.unlocksTechnologies === "undefined" : typeof technology.unlocksTechnologies === "undefined" ? false : this.unlocksTechnologies.every((tech) => technology.unlocksTechnologies.some((otherTech) => tech.strictlyEquals(otherTech))));
  }
};

// src/classes/buildings/Decoration.ts
var Decoration = class _Decoration extends Base {
  //// Static Properties
  /**
   * All the decorations that has been created.
   * @readonly
  */
  static decorations = [];
  /**
   * The description of all decoration builds.
   * @readonly
   */
  static get description() {
    return "Purely decorative, make your factory one of a kind.";
  }
  /**
   * Constructs a new {@link Decoration `Decoration`} object.
   * @param decoration A {@link Decoration `Decoration`} object or decoration parameters.
   */
  constructor(paramsOrDecoration) {
    super(paramsOrDecoration);
    this.image = paramsOrDecoration.image ?? this.image + `decorations/${this.name}.png`;
    _Decoration.decorations.push(this);
  }
};

// src/classes/buildings/tiers/speed-tier/SpeedTier.ts
var SpeedTier = class extends Tier {
  //// Object Properties
  /** The speed at this tier. */
  speed;
  /**
   * Constructs a new {@link SpeedTier `SpeedTier`} object.
   * @param paramsOrSpeedTier A {@link SpeedTier `SpeedTier`} object or speed tier parameters.
   */
  constructor(paramsOrSpeedTier) {
    super(paramsOrSpeedTier);
    this.speed = paramsOrSpeedTier.speed;
  }
  //// Object Methods
  /**
   * These are similarities between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods.
   * @param speedTier The other {@link SpeedTier `SpeedTier`} object.
   * @returns `true` if both {@link SpeedTier `SpeedTier`} objects are the equal in the properties that are similar between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods, `false` otherwise.
   */
  similarEquals(speedTier) {
    return this.speed === speedTier.speed;
  }
  /**
   * Determine if this {@link SpeedTier `SpeedTier`} object is equal to another {@link SpeedTier `SpeedTier`} object.
   * @param speedTier The other {@link SpeedTier `SpeedTier`} object.
   * @returns `true` if both {@link SpeedTier `SpeedTier`} objects are equal, `false` otherwise.
   */
  equals(speedTier) {
    return super.equals(speedTier) && this.similarEquals(speedTier);
  }
  /**
   * Determine if this {@link SpeedTier `SpeedTier`} object is strictly equal to another {@link SpeedTier `SpeedTier`} object.
   * @param speedTier The other {@link SpeedTier `SpeedTier`} object.
   * @returns `true` if both {@link SpeedTier `SpeedTier`} objects are strictly equal, `false` otherwise.
   */
  strictlyEquals(speedTier) {
    return super.strictlyEquals(speedTier) && this.similarEquals(speedTier);
  }
};

// src/classes/buildings/tiers/speed-tier/SpeedTiers.ts
var SpeedTiers = class _SpeedTiers extends Tiers {
  get maxTier() {
    return this.tiers[this.maxTierNum];
  }
  get minTier() {
    return this.tiers[this.minTierNum];
  }
  constructor(paramsOrStorageTiers, passByReference = true) {
    super(paramsOrStorageTiers, passByReference);
    if (paramsOrStorageTiers instanceof _SpeedTiers) {
      if (passByReference) this.tiers = paramsOrStorageTiers.tiers;
      else for (const tier in this.tiers) this.tiers[tier] = new SpeedTier(paramsOrStorageTiers.tiers[tier]);
    } else {
      for (const tier in this.tiers) {
        const t = paramsOrStorageTiers[tier];
        if (t instanceof SpeedTier) this.tiers[tier] = passByReference ? t : new SpeedTier(t);
        else this.tiers[tier] = new SpeedTier(t);
      }
    }
  }
  //// Object Methods
  /**
   * These are similarities between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods.
   * @param speedTiers The other {@link SpeedTiers `StorageTiers`} object.
   * @returns `true` if both {@link SpeedTiers `StorageTiers`} objects are the equal in the things that are similar between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods, `false` otherwise.
   */
  similarEquals(speedTiers) {
    for (const tier in this.tiers) if (this.tiers[tier].speed !== speedTiers.tiers[tier].speed) return false;
    return true;
  }
  /**
   * Determine if this {@link SpeedTiers `StorageTiers`} object is equal to another {@link SpeedTiers `StorageTiers`} object.
   * @param speedTiers The other {@link SpeedTiers `StorageTiers`} object.
   * @returns `true` if both {@link SpeedTiers `StorageTiers`} objects are equal, `false` otherwise.
   */
  equals(speedTiers) {
    return super.equals(speedTiers) && this.similarEquals(speedTiers);
  }
  /**
   * Determine if this {@link SpeedTiers `StorageTiers`} object is strictly equal to another {@link SpeedTiers `StorageTiers`} object.
   * @param speedTiers The other {@link SpeedTiers `StorageTiers`} object.
   * @returns `true` if both {@link SpeedTiers `StorageTiers`} objects are strictly equal, `false` otherwise.
   */
  strictlyEquals(speedTiers) {
    return super.strictlyEquals(speedTiers) && this.similarEquals(speedTiers);
  }
};

// src/classes/buildings/tiers/speed-tier/TransportationSpeed.ts
var TransportationSpeed = class _TransportationSpeed extends SpeedTiers {
  //// Static properties
  /**
   * All the transportations that has been created.
   * @readonly
   */
  static transportationSpeeds = [];
  //// Object Properties
  /** The name of the transportation. */
  name;
  /** The price of the first tier of the transportation. */
  price;
  /** The description of the transportation. */
  description;
  constructor(paramsOrTransportation, passByReference = true) {
    super(paramsOrTransportation.tiers, passByReference);
    this.name = paramsOrTransportation.name;
    this.price = this.tiers[this.minTierNum].price;
    this.description = paramsOrTransportation.description;
    if (paramsOrTransportation instanceof _TransportationSpeed) for (const tier in this.tiers) this.tiers[tier].image = paramsOrTransportation.tiers[tier].image ?? this.tiers[tier].image + `transportations/${this.name === "Robotic Arm" ? `${this.name} Tier ${tier}` : this.name}.png`;
    else {
      if (paramsOrTransportation.tiers instanceof SpeedTiers) for (const tier in this.tiers) this.tiers[tier].image = paramsOrTransportation.tiers.tiers[tier].image ?? this.tiers[tier].image + `transportations/${this.name === "Robotic Arm" ? `${this.name} Tier ${tier}` : this.name}.png`;
      else for (const tier in this.tiers) this.tiers[tier].image = paramsOrTransportation.tiers[tier].image ?? this.tiers[tier].image + `transportations/${this.name === "Robotic Arm" ? `${this.name} Tier ${tier}` : this.name}.png`;
    }
    _TransportationSpeed.transportationSpeeds.push(this);
  }
  //// Object Methods
  /**
   * These are similarities between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods.
   * @param transportationSpeed The other {@link TransportationSpeed `TransportationSpeed`} object.
   * @returns `true` if both {@link TransportationSpeed `TransportationSpeed`} objects are the equal in the things that are similar between the {@link equals `equals`} and {@link strictlyEquals `strictlyEquals`} methods, `false` otherwise.
   */
  similarEquals(transportationSpeed) {
    return this.name === transportationSpeed.name && this.price.equals(transportationSpeed.price) && this.description === transportationSpeed.description;
  }
  /**
   * Determine if this {@link TransportationSpeed `TransportationSpeed`} object is equal to another {@link TransportationSpeed `TransportationSpeed`} object.
   * @param transportationSpeed The other {@link TransportationSpeed `TransportationSpeed`} object.
   * @returns `true` if both storages are the equal, `false` otherwise.
   */
  equals(transportationSpeed) {
    return super.equals(transportationSpeed) && this.similarEquals(transportationSpeed);
  }
  /**
   * Determine if this {@link TransportationSpeed `TransportationSpeed`} object is strictly equal to another {@link TransportationSpeed `TransportationSpeed`} object.
   * @param transportationSpeed speed The other {@link TransportationSpeed `TransportationSpeed`} object.
   * @returns `true` if both storages are the strictly equal, `false` otherwise.
   */
  strictlyEquals(transportationSpeed) {
    return super.strictlyEquals(transportationSpeed) && this.similarEquals(transportationSpeed);
  }
};

// src/objects/currencies.ts
var currencies_exports = {};
__export(currencies_exports, {
  gem: () => gem,
  gold: () => gold
});

// src/classes/Currency.ts
var Currency = class _Currency extends Base {
  //// Static Properties
  /**
   * All the currencies that has been created.
   * @readonly
  */
  static currencies = [];
  /**
   * Constructs a new {@link Currency `Currency`} object.
   * @param currency A {@link Currency `Currency`} object or currency parameters.
   */
  constructor(paramsOrCurrency) {
    super(paramsOrCurrency);
    this.image = paramsOrCurrency.image ?? this.image + `currencies/${this.name}.png`;
    _Currency.currencies.push(this);
  }
};

// src/objects/currencies.ts
var gold = new Currency({
  name: "Gold",
  price: {
    gold: 1,
    gems: null
  }
});
var gem = new Currency({
  name: "Gem",
  price: {
    gold: null,
    gems: 1
  }
});

// src/objects/items.ts
var items_exports = {};
__export(items_exports, {
  atomicLocator: () => atomicLocator,
  battery: () => battery,
  carbonFiber: () => carbonFiber,
  coal: () => coal,
  computer: () => computer,
  concrete: () => concrete,
  condenserLens: () => condenserLens,
  copperIngot: () => copperIngot,
  copperOre: () => copperOre,
  copperWire: () => copperWire,
  coupler: () => coupler,
  earthToken: () => earthToken,
  electricMotor: () => electricMotor,
  electromagnet: () => electromagnet,
  electronMicroscope: () => electronMicroscope,
  emptyFuelCell: () => emptyFuelCell,
  energyCube: () => energyCube,
  enrichedUranium: () => enrichedUranium,
  gemApple: () => gemApple,
  glass: () => glass,
  graphite: () => graphite,
  gyroscope: () => gyroscope,
  heatSink: () => heatSink,
  industrialFrame: () => industrialFrame,
  ironGear: () => ironGear,
  ironIngot: () => ironIngot,
  ironOre: () => ironOre,
  ironPlating: () => ironPlating,
  logicCircuit: () => logicCircuit,
  magneticFieldGenerator: () => magneticFieldGenerator,
  matterCompressor: () => matterCompressor,
  matterDuplicator: () => matterDuplicator,
  metalFrame: () => metalFrame,
  nanoWire: () => nanoWire,
  nuclearFuelCell: () => nuclearFuelCell,
  particleGlue: () => particleGlue,
  quantumEntangler: () => quantumEntangler,
  rotor: () => rotor,
  sand: () => sand,
  silicon: () => silicon,
  stabilizer: () => stabilizer,
  steel: () => steel,
  steelRod: () => steelRod,
  stone: () => stone,
  superComputer: () => superComputer,
  tank: () => tank,
  tungstenCarbide: () => tungstenCarbide,
  tungstenOre: () => tungstenOre,
  turbocharger: () => turbocharger,
  uraniumOre: () => uraniumOre,
  wolframite: () => wolframite,
  woodFrame: () => woodFrame,
  woodLog: () => woodLog,
  woodPlank: () => woodPlank
});
var woodLog = new Item({
  name: "Wood Log",
  price: { gold: 1 },
  factory: extractor,
  resourcesNeeded: [],
  outputPerMin: 7.5
});
var stone = new Item({
  name: "Stone",
  price: { gold: 1 },
  factory: extractor,
  resourcesNeeded: [],
  outputPerMin: 7.5
});
var ironOre = new Item({
  name: "Iron Ore",
  price: { gold: 1 },
  factory: extractor,
  resourcesNeeded: [],
  outputPerMin: 7.5
});
var copperOre = new Item({
  name: "Copper Ore",
  price: { gold: 1 },
  factory: extractor,
  resourcesNeeded: [],
  outputPerMin: 7.5
});
var coal = new Item({
  name: "Coal",
  price: { gold: 1 },
  factory: extractor,
  resourcesNeeded: [],
  outputPerMin: 7.5
});
coalPowerPlant.input = {
  item: coal,
  amount: 5,
  inputPerMin: 20
};
var wolframite = new Item({
  name: "Wolframite",
  price: { gold: 1 },
  factory: extractor,
  resourcesNeeded: [],
  outputPerMin: 7.5
});
var woodPlank = new Item({
  name: "Wood Plank",
  price: { gold: 1 },
  factory: workshop,
  resourcesNeeded: [
    {
      item: woodLog,
      amount: 1
    }
  ],
  outputPerMin: 15
});
var sand = new Item({
  name: "Sand",
  price: { gold: 1 },
  factory: workshop,
  resourcesNeeded: [
    {
      item: stone,
      amount: 1
    }
  ],
  outputPerMin: 40
});
var ironIngot = new Item({
  name: "Iron Ingot",
  price: { gold: 1 },
  factory: furnace,
  resourcesNeeded: [
    {
      item: ironOre,
      amount: 1
    }
  ],
  outputPerMin: 30
});
var copperIngot = new Item({
  name: "Copper Ingot",
  price: { gold: 1 },
  factory: furnace,
  resourcesNeeded: [
    {
      item: copperOre,
      amount: 1
    }
  ],
  outputPerMin: 30
});
var tungstenOre = new Item({
  name: "Tungsten Ore",
  price: { gold: 5 },
  factory: furnace,
  resourcesNeeded: [
    {
      item: wolframite,
      amount: 5
    }
  ],
  outputPerMin: 24
});
var graphite = new Item({
  name: "Graphite",
  price: { gold: 6 },
  factory: forge,
  resourcesNeeded: [
    {
      item: coal,
      amount: 3
    },
    {
      item: woodLog,
      amount: 3
    }
  ],
  outputPerMin: 15
});
var woodFrame = new Item({
  name: "Wood Frame",
  price: { gold: 1 },
  factory: workshop,
  resourcesNeeded: [
    {
      item: woodPlank,
      amount: 4
    }
  ],
  outputPerMin: 7.5
});
var copperWire = new Item({
  name: "Copper Wire",
  price: { gold: 2 },
  factory: workshop,
  resourcesNeeded: [
    {
      item: copperIngot,
      amount: 3
    }
  ],
  outputPerMin: 30,
  outputAmount: 2
});
var heatSink = new Item({
  name: "Heat Sink",
  price: { gold: 5 },
  factory: workshop,
  resourcesNeeded: [
    {
      item: copperIngot,
      amount: 5
    }
  ],
  outputPerMin: 10
});
var ironGear = new Item({
  name: "Iron Gear",
  price: { gold: 2 },
  factory: workshop,
  resourcesNeeded: [
    {
      item: ironIngot,
      amount: 2
    }
  ],
  outputPerMin: 15
});
var ironPlating = new Item({
  name: "Iron Plating",
  price: { gold: 2 },
  factory: workshop,
  resourcesNeeded: [
    {
      item: ironIngot,
      amount: 4
    }
  ],
  outputPerMin: 20,
  outputAmount: 2
});
var steel = new Item({
  name: "Steel",
  price: { gold: 12 },
  factory: forge,
  resourcesNeeded: [
    {
      item: graphite,
      amount: 1
    },
    {
      item: ironOre,
      amount: 6
    }
  ],
  outputPerMin: 7.5
});
var steelRod = new Item({
  name: "Steel Rod",
  price: { gold: 35 },
  factory: workshop,
  resourcesNeeded: [
    {
      item: steel,
      amount: 3
    }
  ],
  outputPerMin: 15
});
var glass = new Item({
  name: "Glass",
  price: { gold: 4 },
  factory: furnace,
  resourcesNeeded: [
    {
      item: sand,
      amount: 4
    }
  ],
  outputPerMin: 10
});
var condenserLens = new Item({
  name: "Condenser Lens",
  price: { gold: 12 },
  factory: workshop,
  resourcesNeeded: [
    {
      item: glass,
      amount: 3
    }
  ],
  outputPerMin: 40
});
var carbonFiber = new Item({
  name: "Carbon Fiber",
  price: { gold: 24 },
  factory: workshop,
  resourcesNeeded: [
    {
      item: graphite,
      amount: 4
    }
  ],
  outputPerMin: 7.5
});
var tungstenCarbide = new Item({
  name: "Tungsten Carbide",
  price: { gold: 16 },
  factory: forge,
  resourcesNeeded: [
    {
      item: tungstenOre,
      amount: 2
    },
    {
      item: graphite,
      amount: 1
    }
  ],
  outputPerMin: 12
});
var coupler = new Item({
  name: "Coupler",
  price: { gold: 16 },
  factory: workshop,
  resourcesNeeded: [
    {
      item: tungstenCarbide,
      amount: 1
    }
  ],
  outputPerMin: 6
});
var concrete = new Item({
  name: "Concrete",
  price: { gold: 40 },
  factory: forge,
  resourcesNeeded: [
    {
      item: sand,
      amount: 10
    },
    {
      item: steelRod,
      amount: 1
    }
  ],
  outputPerMin: 7.5
});
var metalFrame = new Item({
  name: "Metal Frame",
  price: { gold: 12 },
  factory: machineShop,
  resourcesNeeded: [
    {
      item: woodFrame,
      amount: 1
    },
    {
      item: ironPlating,
      amount: 4
    }
  ],
  outputPerMin: 5
});
var industrialFrame = new Item({
  name: "Industrial Frame",
  price: { gold: 400 },
  factory: industrialFactory,
  resourcesNeeded: [
    {
      item: concrete,
      amount: 6
    },
    {
      item: metalFrame,
      amount: 2
    },
    {
      item: tungstenCarbide,
      amount: 8
    }
  ],
  outputPerMin: 3
});
var silicon = new Item({
  name: "Silicon",
  price: { gold: 2 },
  factory: furnace,
  resourcesNeeded: [
    {
      item: sand,
      amount: 2
    }
  ],
  outputPerMin: 20
});
var logicCircuit = new Item({
  name: "Logic Circuit",
  price: { gold: 10 },
  factory: machineShop,
  resourcesNeeded: [
    {
      item: copperWire,
      amount: 3
    },
    {
      item: silicon,
      amount: 2
    }
  ],
  outputPerMin: 10
});
var nanoWire = new Item({
  name: "Nano Wire",
  price: { gold: 60 },
  factory: machineShop,
  resourcesNeeded: [
    {
      item: carbonFiber,
      amount: 2
    },
    {
      item: glass,
      amount: 4
    }
  ],
  outputPerMin: 5
});
var turbocharger = new Item({
  name: "Turbocharger",
  price: { gold: 250 },
  factory: manufacturer,
  resourcesNeeded: [
    {
      item: ironGear,
      amount: 8
    },
    {
      item: logicCircuit,
      amount: 4
    },
    {
      item: nanoWire,
      amount: 2
    },
    {
      item: coupler,
      amount: 4
    }
  ],
  outputPerMin: 4
});
var electromagnet = new Item({
  name: "Electromagnet",
  price: { gold: 14 },
  factory: machineShop,
  resourcesNeeded: [
    {
      item: copperWire,
      amount: 6
    },
    {
      item: ironIngot,
      amount: 2
    }
  ],
  outputPerMin: 10
});
var battery = new Item({
  name: "Battery",
  price: { gold: 150 },
  factory: machineShop,
  resourcesNeeded: [
    {
      item: electromagnet,
      amount: 8
    },
    {
      item: graphite,
      amount: 8
    }
  ],
  outputPerMin: 2.5
});
var rotor = new Item({
  name: "Rotor",
  price: { gold: 40 },
  factory: machineShop,
  resourcesNeeded: [
    {
      item: steelRod,
      amount: 1
    },
    {
      item: ironPlating,
      amount: 2
    }
  ],
  outputPerMin: 10
});
var electricMotor = new Item({
  name: "Electric Motor",
  price: { gold: 250 },
  factory: industrialFactory,
  resourcesNeeded: [
    {
      item: battery,
      amount: 1
    },
    {
      item: ironGear,
      amount: 4
    },
    {
      item: rotor,
      amount: 2
    }
  ],
  outputPerMin: 3
});
var tank = new Item({
  name: "Tank",
  price: { gold: 250 },
  factory: industrialFactory,
  resourcesNeeded: [
    {
      item: glass,
      amount: 2
    },
    {
      item: concrete,
      amount: 4
    },
    {
      item: tungstenCarbide,
      amount: 4
    }
  ],
  outputPerMin: 6
});
var matterCompressor = new Item({
  name: "Matter Compressor",
  price: { gold: 1500 },
  factory: manufacturer,
  resourcesNeeded: [
    {
      item: industrialFrame,
      amount: 1
    },
    {
      item: turbocharger,
      amount: 2
    },
    {
      item: electricMotor,
      amount: 2
    },
    {
      item: tank,
      amount: 1
    }
  ],
  outputPerMin: 2
});
var particleGlue = new Item({
  name: "Particle Glue",
  price: { gold: 150 },
  factory: workshop,
  resourcesNeeded: [
    {
      item: matterCompressor,
      amount: 1
    }
  ],
  outputPerMin: 20,
  outputAmount: 10
});
var gyroscope = new Item({
  name: "Gyroscope",
  price: { gold: 100 },
  factory: machineShop,
  resourcesNeeded: [
    {
      item: copperWire,
      amount: 12
    },
    {
      item: rotor,
      amount: 2
    }
  ],
  outputPerMin: 5
});
var energyCube = new Item({
  name: "Energy Cube",
  price: { gold: 700 },
  factory: machineShop,
  resourcesNeeded: [
    {
      item: battery,
      amount: 2
    },
    {
      item: industrialFrame,
      amount: 1
    }
  ],
  outputPerMin: 2
});
var computer = new Item({
  name: "Computer",
  price: { gold: 60 },
  factory: industrialFactory,
  resourcesNeeded: [
    {
      item: metalFrame,
      amount: 1
    },
    {
      item: heatSink,
      amount: 3
    },
    {
      item: logicCircuit,
      amount: 3
    }
  ],
  outputPerMin: 7.5
});
var stabilizer = new Item({
  name: "Stabilizer",
  price: { gold: 500 },
  factory: industrialFactory,
  resourcesNeeded: [
    {
      item: computer,
      amount: 1
    },
    {
      item: electricMotor,
      amount: 1
    },
    {
      item: gyroscope,
      amount: 2
    }
  ],
  outputPerMin: 2.5
});
var magneticFieldGenerator = new Item({
  name: "Magnetic Field Generator",
  price: { gold: 1500 },
  factory: manufacturer,
  resourcesNeeded: [
    {
      item: stabilizer,
      amount: 1
    },
    {
      item: industrialFrame,
      amount: 1
    },
    {
      item: electromagnet,
      amount: 10
    },
    {
      item: nanoWire,
      amount: 10
    }
  ],
  outputPerMin: 1.5
});
var quantumEntangler = new Item({
  name: "Quantum Entangler",
  price: { gold: 2500 },
  factory: machineShop,
  resourcesNeeded: [
    {
      item: magneticFieldGenerator,
      amount: 1
    },
    {
      item: stabilizer,
      amount: 2
    }
  ],
  outputPerMin: 1
});
var electronMicroscope = new Item({
  name: "Electron Microscope",
  price: { gold: 300 },
  factory: manufacturer,
  resourcesNeeded: [
    {
      item: nanoWire,
      amount: 2
    },
    {
      item: electromagnet,
      amount: 8
    },
    {
      item: condenserLens,
      amount: 4
    },
    {
      item: metalFrame,
      amount: 2
    }
  ],
  outputPerMin: 2.5
});
var superComputer = new Item({
  name: "Super Computer",
  price: { gold: 500 },
  factory: manufacturer,
  resourcesNeeded: [
    {
      item: computer,
      amount: 2
    },
    {
      item: heatSink,
      amount: 8
    },
    {
      item: turbocharger,
      amount: 1
    },
    {
      item: coupler,
      amount: 8
    }
  ],
  outputPerMin: 2
});
var atomicLocator = new Item({
  name: "Atomic Locator",
  price: { gold: 2500 },
  factory: manufacturer,
  resourcesNeeded: [
    {
      item: superComputer,
      amount: 2
    },
    {
      item: electronMicroscope,
      amount: 2
    },
    {
      item: concrete,
      amount: 24
    },
    {
      item: copperWire,
      amount: 50
    }
  ],
  outputPerMin: 2
});
var matterDuplicator = new Item({
  name: "Matter Duplicator",
  price: { gold: 32e3 },
  factory: manufacturer,
  resourcesNeeded: [
    {
      item: atomicLocator,
      amount: 4
    },
    {
      item: quantumEntangler,
      amount: 2
    },
    {
      item: energyCube,
      amount: 5
    },
    {
      item: particleGlue,
      amount: 100
    }
  ],
  outputPerMin: 2 / 3
});
var earthToken = new Item({
  name: "Earth Token",
  price: { gold: 32e3 },
  factory: earthTeleporter,
  resourcesNeeded: [
    {
      item: matterDuplicator,
      amount: 1
    }
  ],
  outputPerMin: 60 / 42
});
var uraniumOre = new Item({
  name: "Uranium Ore",
  price: { gold: 10 },
  factory: uraniumExtractor,
  resourcesNeeded: [],
  outputPerMin: 10
});
var enrichedUranium = new Item({
  name: "Enriched Uranium",
  price: { gold: 1e3 },
  factory: furnace,
  resourcesNeeded: [
    {
      item: uraniumOre,
      amount: 30
    }
  ],
  outputPerMin: 1
});
var emptyFuelCell = new Item({
  name: "Empty Fuel Cell",
  price: { gold: 70 },
  factory: machineShop,
  resourcesNeeded: [
    {
      item: tungstenCarbide,
      amount: 3
    },
    {
      item: glass,
      amount: 5
    }
  ],
  outputPerMin: 4
});
var nuclearFuelCell = new Item({
  name: "Nuclear Fuel Cell",
  price: { gold: 1200 },
  factory: industrialFactory,
  resourcesNeeded: [
    {
      item: emptyFuelCell,
      amount: 1
    },
    {
      item: steelRod,
      amount: 1
    },
    {
      item: enrichedUranium,
      amount: 1
    }
  ],
  outputPerMin: 2
});
nuclearPowerPlant.input = {
  item: nuclearFuelCell,
  amount: 1,
  inputPerMin: 1
};
var gemApple = new Item({
  name: "Gem Apple",
  price: { gems: 1 },
  factory: gemTree,
  resourcesNeeded: [],
  outputPerMin: 0.2
});

// src/objects/technologies.ts
var technologies_exports = {};
__export(technologies_exports, {
  advancedBelts: () => advancedBelts,
  atomicLocator: () => atomicLocator2,
  batteries: () => batteries,
  circuitStatue: () => circuitStatue2,
  circuitry: () => circuitry,
  coalPower: () => coalPower,
  computers: () => computers,
  concrete: () => concrete2,
  decorativeRocks: () => decorativeRocks,
  decorativeTrees: () => decorativeTrees,
  decorativeWalls: () => decorativeWalls,
  earthStatue: () => earthStatue2,
  earthTeleporter: () => earthTeleporter2,
  energyCube: () => energyCube2,
  extractorUpgrade1: () => extractorUpgrade1,
  extractorUpgrade2: () => extractorUpgrade2,
  extractorUpgrade3: () => extractorUpgrade3,
  extractorUpgrade4: () => extractorUpgrade4,
  factoryDecorations: () => factoryDecorations,
  festiveDecorations: () => festiveDecorations,
  forgeUpgrade1: () => forgeUpgrade1,
  forgeUpgrade2: () => forgeUpgrade2,
  forgeUpgrade3: () => forgeUpgrade3,
  furnaceUpgrade1: () => furnaceUpgrade1,
  furnaceUpgrade2: () => furnaceUpgrade2,
  furnaceUpgrade3: () => furnaceUpgrade3,
  gearStatue: () => gearStatue2,
  gemTree: () => gemTree2,
  goldVault: () => goldVault2,
  industrialFactoryUpgrade1: () => industrialFactoryUpgrade1,
  industrialFactoryUpgrade2: () => industrialFactoryUpgrade2,
  industrialFactoryUpgrade3: () => industrialFactoryUpgrade3,
  industrialFrame: () => industrialFrame2,
  machineShopUpgrade1: () => machineShopUpgrade1,
  machineShopUpgrade2: () => machineShopUpgrade2,
  machineShopUpgrade3: () => machineShopUpgrade3,
  magnets: () => magnets,
  manufacturerUpgrade1: () => manufacturerUpgrade1,
  manufacturerUpgrade2: () => manufacturerUpgrade2,
  manufacturerUpgrade3: () => manufacturerUpgrade3,
  matterCompressor: () => matterCompressor2,
  matterDuplication: () => matterDuplication,
  metalFrames: () => metalFrames,
  metalWorking: () => metalWorking,
  microscopes: () => microscopes,
  motors: () => motors,
  nanoWire: () => nanoWire2,
  nuclearPower: () => nuclearPower,
  particleStatue: () => particleStatue2,
  quantumEntanglement: () => quantumEntanglement,
  roboticArmUpgrade1: () => roboticArmUpgrade1,
  roboticArmUpgrade2: () => roboticArmUpgrade2,
  roboticArmUpgrade3: () => roboticArmUpgrade3,
  robotics: () => robotics,
  sand: () => sand2,
  smelting: () => smelting,
  speedUpgradeI: () => speedUpgradeI,
  speedUpgradeII: () => speedUpgradeII,
  speedUpgradeIII: () => speedUpgradeIII,
  speedUpgradeIV: () => speedUpgradeIV,
  speedUpgradeIX: () => speedUpgradeIX,
  speedUpgradeV: () => speedUpgradeV,
  speedUpgradeVI: () => speedUpgradeVI,
  speedUpgradeVII: () => speedUpgradeVII,
  speedUpgradeVIII: () => speedUpgradeVIII,
  speedUpgradeX: () => speedUpgradeX,
  speedUpgradeXI: () => speedUpgradeXI,
  speedUpgradeXII: () => speedUpgradeXII,
  stabilizer: () => stabilizer2,
  startingTech: () => startingTech,
  steelProduction: () => steelProduction,
  steelStatue: () => steelStatue2,
  storage: () => storage2,
  storageUpgrade1: () => storageUpgrade1,
  storageUpgrade2: () => storageUpgrade2,
  storageUpgrade3: () => storageUpgrade3,
  superComputers: () => superComputers,
  tungstenProduction: () => tungstenProduction,
  turbocharger: () => turbocharger2,
  undergroundBeltUpgrade1: () => undergroundBeltUpgrade1,
  undergroundBeltUpgrade2: () => undergroundBeltUpgrade2,
  uraniumProcessing: () => uraniumProcessing,
  woodFrame: () => woodFrame2,
  woodPlanks: () => woodPlanks,
  workshopUpgrade1: () => workshopUpgrade1,
  workshopUpgrade2: () => workshopUpgrade2,
  workshopUpgrade3: () => workshopUpgrade3
});

// src/objects/buildings/decorations.ts
var decorations_exports = {};
__export(decorations_exports, {
  barrel: () => barrel,
  birchTree: () => birchTree,
  brokenPipe: () => brokenPipe,
  circuitStatue: () => circuitStatue,
  concreteWall: () => concreteWall,
  earthStatue: () => earthStatue,
  fence: () => fence,
  firTree: () => firTree,
  gearStatue: () => gearStatue,
  hazardCone: () => hazardCone,
  palmTree: () => palmTree,
  particleStatue: () => particleStatue,
  prettyTree: () => prettyTree,
  pumpkin: () => pumpkin,
  rock1: () => rock1,
  rock2: () => rock2,
  rock3: () => rock3,
  snowman: () => snowman,
  steelStatue: () => steelStatue,
  woodenCrate: () => woodenCrate
});
var fence = new Decoration({
  name: "Fence",
  price: { gems: 2 }
});
var concreteWall = new Decoration({
  name: "Concrete Wall",
  price: { gems: 2 }
});
var prettyTree = new Decoration({
  name: "Pretty Tree",
  price: { gems: 10 }
});
var firTree = new Decoration({
  name: "Fir Tree",
  price: { gems: 10 }
});
var birchTree = new Decoration({
  name: "Birch Tree",
  price: { gems: 10 }
});
var palmTree = new Decoration({
  name: "Palm Tree",
  price: { gems: 10 }
});
var rock1 = new Decoration({
  name: "Rock",
  price: { gems: 10 },
  image: "./resources/decorations/Rock 1.png"
});
var rock2 = new Decoration({
  name: "Rock",
  price: { gems: 10 },
  image: "./resources/decorations/Rock 2.png"
});
var rock3 = new Decoration({
  name: "Rock",
  price: { gems: 10 },
  image: "./resources/decorations/Rock 3.png"
});
var hazardCone = new Decoration({
  name: "Hazard Cone",
  price: { gems: 20 }
});
var brokenPipe = new Decoration({
  name: "Broken Pipe",
  price: { gems: 20 }
});
var barrel = new Decoration({
  name: "Barrel",
  price: { gems: 20 }
});
var woodenCrate = new Decoration({
  name: "Wooden Crate",
  price: { gems: 20 }
});
var gearStatue = new Decoration({
  name: "Gear Statue",
  price: { gems: 20 }
});
var circuitStatue = new Decoration({
  name: "Circuit Statue",
  price: { gems: 40 }
});
var steelStatue = new Decoration({
  name: "Steel Statue",
  price: { gems: 60 }
});
var particleStatue = new Decoration({
  name: "Particle Statue",
  price: { gems: 80 }
});
var earthStatue = new Decoration({
  name: "Earth Statue",
  price: { gems: 100 }
});
var pumpkin = new Decoration({
  name: "Pumpkin",
  price: { gems: 30 }
});
var snowman = new Decoration({
  name: "Snowman",
  price: { gems: 30 }
});

// src/objects/buildings/storages.ts
var storages_exports = {};
__export(storages_exports, {
  storage: () => storage
});
var storage = new Storage({
  name: "Storage",
  description: "Storage buildings can only store one type of item at a time. Max capacity can be upgraded.",
  tiers: {
    1: {
      price: { gold: 500 },
      storage: 250
    },
    2: {
      price: { gold: 500 },
      storage: 500
    },
    3: {
      price: { gold: 1e3 },
      storage: 1e3
    },
    4: {
      price: { gold: 2e3 },
      storage: 2e3
    }
  }
});

// src/objects/buildings/item-collectors.ts
var item_collectors_exports = {};
__export(item_collectors_exports, {
  goldVault: () => goldVault,
  researchLab: () => researchLab
});
var researchLab = new ItemCollector({
  name: "Research Lab",
  price: { gold: 1e3 },
  description: "Item sent into the Research Lab can be used to unlock technology. Items not used for research will give you gold instead.",
  limit: 3,
  inputs: 4
});
var goldVault = new ItemCollector({
  name: "Gold Vault",
  price: { gold: 100 },
  description: "Item sent into the Gold Vault converted into gold that can be used to build your factory.",
  limit: 3,
  inputs: 4
});

// src/objects/buildings/transportations.ts
var transportations_exports = {};
__export(transportations_exports, {
  BeltSpeeds: () => BeltSpeeds,
  UndergroundBeltDistances: () => UndergroundBeltDistances,
  beltSplitter: () => beltSplitter,
  conveyorBelt: () => conveyorBelt,
  roboticArm: () => roboticArm,
  undergroundBelt: () => undergroundBelt
});
var conveyorBelt = new TransportationSpeed({
  name: "Conveyor Belt",
  description: "Conveyor Belts transport items around the world. Connect to belts, extractors, factories, and other buildings.",
  tiers: {
    0: {
      price: { gold: 4 },
      speed: 150
    },
    1: { speed: 165 },
    2: { speed: 180 },
    3: { speed: 195 },
    4: { speed: 210 },
    5: { speed: 240 },
    6: { speed: 270 },
    7: { speed: 300 },
    8: { speed: 330 },
    9: { speed: 375 },
    10: { speed: 420 },
    11: { speed: 450 },
    12: { speed: 480 }
  }
});
var BeltSpeeds = Object.values(conveyorBelt.tiers).map((tier) => tier.speed);
var undergroundBelt = new TransportationDistance({
  name: "Underground Belt",
  description: "Underground Belts transport items underground, allowing them to avoid obstacles. Connect to another underground belt entrance / exist with up to 4 tiles of space between them.",
  tiers: {
    0: {
      price: { gold: 10 },
      distance: 4
    },
    1: { distance: 5 },
    2: { distance: 6 }
  }
});
var UndergroundBeltDistances = Object.values(undergroundBelt.tiers).map((tier) => tier.distance);
var beltSplitter = new TransportationDistance({
  name: "Belt Splitter",
  description: "Belt Splitters evenly distribute items to connected belts or factories.",
  tiers: {
    0: {
      price: { gold: 10 },
      distance: 1
    }
  }
});
var roboticArm = new TransportationSpeed({
  name: "Robotic Arm",
  description: "Robotic Arms can move items between buildings and belts from any direction. They can also filter for specific items.",
  tiers: {
    1: {
      price: { gold: 250 },
      speed: 1
    },
    2: {
      price: { gold: 800 },
      speed: 2
    },
    3: {
      price: { gold: 2e3 },
      speed: 4
    },
    4: {
      price: { gold: 8e3 },
      speed: 6
    }
  }
});

// src/objects/technologies.ts
var manufacturerUpgrade3 = new Technology({
  name: "Manufacturer Upgrade",
  price: { gems: 452300 },
  unlocksTechnologies: [],
  resourcesNeeded: [
    {
      item: industrialFrame,
      amount: 2e3
    },
    {
      item: superComputer,
      amount: 100
    }
  ],
  unlocks: [
    {
      build: manufacturer,
      tier: 4
    }
  ]
});
var manufacturerUpgrade2 = new Technology({
  name: "Manufacturer Upgrade",
  price: { gems: 106500 },
  unlocksTechnologies: [manufacturerUpgrade3],
  resourcesNeeded: [
    {
      item: industrialFrame,
      amount: 500
    }
  ],
  unlocks: [
    {
      build: manufacturer,
      tier: 3
    }
  ]
});
var manufacturerUpgrade1 = new Technology({
  name: "Manufacturer Upgrade",
  price: { gems: 25100 },
  unlocksTechnologies: [manufacturerUpgrade2],
  resourcesNeeded: [
    {
      item: tungstenCarbide,
      amount: 2e3
    },
    {
      item: steel,
      amount: 1e3
    }
  ],
  unlocks: [
    {
      build: manufacturer,
      tier: 2
    }
  ]
});
var forgeUpgrade3 = new Technology({
  name: "Forge Upgrade",
  price: { gems: 132500 },
  unlocksTechnologies: [],
  resourcesNeeded: [
    {
      item: superComputer,
      amount: 500
    }
  ],
  unlocks: [
    {
      build: forge,
      tier: 4
    }
  ]
});
var forgeUpgrade2 = new Technology({
  name: "Forge Upgrade",
  price: { gems: 27100 },
  unlocksTechnologies: [forgeUpgrade3],
  resourcesNeeded: [
    {
      item: carbonFiber,
      amount: 2e3
    }
  ],
  unlocks: [
    {
      build: forge,
      tier: 3
    }
  ]
});
var forgeUpgrade1 = new Technology({
  name: "Forge Upgrade",
  price: { gems: 9800 },
  unlocksTechnologies: [forgeUpgrade2],
  resourcesNeeded: [
    {
      item: steelRod,
      amount: 500
    }
  ],
  unlocks: [
    {
      build: forge,
      tier: 2
    }
  ]
});
var industrialFactoryUpgrade3 = new Technology({
  name: "Industrial Factory Upgrade",
  price: { gems: 106500 },
  unlocksTechnologies: [],
  resourcesNeeded: [
    {
      item: industrialFrame,
      amount: 500
    }
  ],
  unlocks: [
    {
      build: industrialFactory,
      tier: 4
    }
  ]
});
var industrialFactoryUpgrade2 = new Technology({
  name: "Industrial Factory Upgrade",
  price: { gems: 31400 },
  unlocksTechnologies: [industrialFactoryUpgrade3],
  resourcesNeeded: [
    {
      item: tungstenCarbide,
      amount: 1e3
    },
    {
      item: rotor,
      amount: 1e3
    }
  ],
  unlocks: [
    {
      build: industrialFactory,
      tier: 3
    }
  ]
});
var industrialFactoryUpgrade1 = new Technology({
  name: "Industrial Factory Upgrade",
  price: { gems: 6900 },
  unlocksTechnologies: [industrialFactoryUpgrade2],
  resourcesNeeded: [
    {
      item: steel,
      amount: 1e3
    }
  ],
  unlocks: [
    {
      build: industrialFactory,
      tier: 2
    }
  ]
});
var gemTree2 = new Technology({
  name: "Gem Tree",
  price: { gems: 2400 },
  unlocksTechnologies: [],
  resourcesNeeded: [
    {
      item: woodLog,
      amount: 1e3
    },
    {
      item: stone,
      amount: 1e3
    },
    {
      item: ironOre,
      amount: 1e3
    },
    {
      item: copperOre,
      amount: 1e3
    }
  ],
  unlocks: [
    {
      build: gemTree,
      tier: 1
    }
  ]
});
var earthStatue2 = new Technology({
  name: "Earth Statue",
  price: { gems: null },
  unlocksTechnologies: [],
  resourcesNeeded: [
    {
      item: earthToken,
      amount: 500
    }
  ],
  unlocks: [earthStatue]
});
var particleStatue2 = new Technology({
  name: "Particle Statue",
  price: { gems: null },
  unlocksTechnologies: [earthStatue2],
  resourcesNeeded: [
    {
      item: particleGlue,
      amount: 1e3
    }
  ],
  unlocks: [particleStatue]
});
var steelStatue2 = new Technology({
  name: "Steel Statue",
  price: { gems: null },
  unlocksTechnologies: [particleStatue2],
  resourcesNeeded: [
    {
      item: steel,
      amount: 5e3
    }
  ],
  unlocks: [steelStatue]
});
var circuitStatue2 = new Technology({
  name: "Circuit Statue",
  price: { gems: null },
  unlocksTechnologies: [steelStatue2],
  resourcesNeeded: [
    {
      item: logicCircuit,
      amount: 3e3
    }
  ],
  unlocks: [circuitStatue]
});
var gearStatue2 = new Technology({
  name: "Gear Statue",
  price: { gems: null },
  unlocksTechnologies: [circuitStatue2],
  resourcesNeeded: [
    {
      item: ironGear,
      amount: 2e3
    }
  ],
  unlocks: [gearStatue]
});
var decorativeWalls = new Technology({
  name: "Decorative Walls",
  price: { gems: 1300 },
  unlocksTechnologies: [],
  resourcesNeeded: [
    {
      item: woodPlank,
      amount: 200
    },
    {
      item: concrete,
      amount: 50
    }
  ],
  unlocks: [fence, concreteWall]
});
var factoryDecorations = new Technology({
  name: "Factory Decorations",
  price: { gems: 360 },
  unlocksTechnologies: [decorativeWalls],
  resourcesNeeded: [
    {
      item: copperWire,
      amount: 150
    },
    {
      item: ironGear,
      amount: 150
    }
  ],
  unlocks: [hazardCone, barrel, brokenPipe, woodenCrate]
});
var decorativeRocks = new Technology({
  name: "Decorative Rocks",
  price: { gems: 300 },
  unlocksTechnologies: [factoryDecorations],
  resourcesNeeded: [
    {
      item: stone,
      amount: 500
    }
  ],
  unlocks: [rock1, rock2, rock3]
});
var festiveDecorations = new Technology({
  name: "Festive Decorations",
  price: { gems: 300 },
  unlocksTechnologies: [],
  resourcesNeeded: [
    {
      item: coal,
      amount: 50
    }
  ],
  unlocks: [pumpkin, snowman]
});
var decorativeTrees = new Technology({
  name: "Decorative Trees",
  price: { gems: 120 },
  unlocksTechnologies: [festiveDecorations, decorativeRocks, gearStatue2],
  resourcesNeeded: [
    {
      item: woodFrame,
      amount: 50
    }
  ],
  unlocks: [prettyTree, firTree, birchTree, palmTree]
});
var roboticArmUpgrade3 = new Technology({
  name: "Robotic Arm Upgrade",
  price: { gems: 133500 },
  unlocksTechnologies: [],
  resourcesNeeded: [
    {
      item: electricMotor,
      amount: 5e3
    },
    {
      item: nanoWire,
      amount: 2e3
    }
  ],
  unlocks: [
    {
      build: roboticArm,
      tier: 4
    }
  ]
});
var roboticArmUpgrade2 = new Technology({
  name: "Robotic Arm Upgrade",
  price: { gems: 22300 },
  unlocksTechnologies: [roboticArmUpgrade3],
  resourcesNeeded: [
    {
      item: rotor,
      amount: 1e3
    }
  ],
  unlocks: [
    {
      build: roboticArm,
      tier: 3
    }
  ]
});
var roboticArmUpgrade1 = new Technology({
  name: "Robotic Arm Upgrade",
  price: { gems: 6700 },
  unlocksTechnologies: [roboticArmUpgrade2],
  resourcesNeeded: [
    {
      item: computer,
      amount: 200
    }
  ],
  unlocks: [
    {
      build: roboticArm,
      tier: 2
    }
  ]
});
var robotics = new Technology({
  name: "Robotics",
  price: { gems: 2100 },
  unlocksTechnologies: [gemTree2, roboticArmUpgrade1],
  resourcesNeeded: [
    {
      item: woodPlank,
      amount: 1e3
    },
    {
      item: logicCircuit,
      amount: 2500
    }
  ],
  unlocks: [
    {
      build: roboticArm,
      tier: 1
    }
  ]
});
var machineShopUpgrade3 = new Technology({
  name: "Machine Shop Upgrade",
  price: { gems: 40500 },
  unlocksTechnologies: [],
  resourcesNeeded: [
    {
      item: concrete,
      amount: 1e3
    },
    {
      item: coupler,
      amount: 2e3
    }
  ],
  unlocks: [
    {
      build: machineShop,
      tier: 4
    }
  ]
});
var machineShopUpgrade2 = new Technology({
  name: "Machine Shop Upgrade",
  price: { gems: 16600 },
  unlocksTechnologies: [machineShopUpgrade3],
  resourcesNeeded: [
    {
      item: computer,
      amount: 500
    }
  ],
  unlocks: [
    {
      build: machineShop,
      tier: 3
    }
  ]
});
var machineShopUpgrade1 = new Technology({
  name: "Machine Shop Upgrade",
  price: { gems: 3200 },
  unlocksTechnologies: [machineShopUpgrade2],
  resourcesNeeded: [
    {
      item: ironPlating,
      amount: 2e3
    },
    {
      item: electromagnet,
      amount: 100
    }
  ],
  unlocks: [
    {
      build: machineShop,
      tier: 2
    }
  ]
});
var storageUpgrade3 = new Technology({
  name: "Storage Upgrade",
  price: { gems: 49700 },
  unlocksTechnologies: [],
  resourcesNeeded: [
    {
      item: metalFrame,
      amount: 4e3
    },
    {
      item: concrete,
      amount: 1e3
    }
  ],
  unlocks: [
    {
      build: storage,
      tier: 4
    }
  ]
});
var storageUpgrade2 = new Technology({
  name: "Storage Upgrade",
  price: { gems: 17200 },
  unlocksTechnologies: [storageUpgrade3],
  resourcesNeeded: [
    {
      item: metalFrame,
      amount: 2e3
    },
    {
      item: steel,
      amount: 500
    }
  ],
  unlocks: [
    {
      build: storage,
      tier: 3
    }
  ]
});
var storageUpgrade1 = new Technology({
  name: "Storage Upgrade",
  price: { gems: 6900 },
  unlocksTechnologies: [storageUpgrade2],
  resourcesNeeded: [
    {
      item: metalFrame,
      amount: 1e3
    }
  ],
  unlocks: [
    {
      build: storage,
      tier: 2
    }
  ]
});
var storage2 = new Technology({
  name: "Storage",
  price: { gems: 1100 },
  unlocksTechnologies: [storageUpgrade1],
  resourcesNeeded: [
    {
      item: metalFrame,
      amount: 150
    }
  ],
  unlocks: [
    {
      build: storage,
      tier: 1
    }
  ]
});
var furnaceUpgrade3 = new Technology({
  name: "Furnace Upgrade",
  price: { gems: 33200 },
  unlocksTechnologies: [],
  resourcesNeeded: [
    {
      item: nanoWire,
      amount: 1e3
    }
  ],
  unlocks: [
    {
      build: furnace,
      tier: 4
    }
  ]
});
var furnaceUpgrade2 = new Technology({
  name: "Furnace Upgrade",
  price: { gems: 14600 },
  unlocksTechnologies: [furnaceUpgrade3],
  resourcesNeeded: [
    {
      item: heatSink,
      amount: 5e3
    }
  ],
  unlocks: [
    {
      build: furnace,
      tier: 3
    }
  ]
});
var furnaceUpgrade1 = new Technology({
  name: "Furnace Upgrade",
  price: { gems: 4e3 },
  unlocksTechnologies: [furnaceUpgrade2],
  resourcesNeeded: [
    {
      item: electromagnet,
      amount: 500
    }
  ],
  unlocks: [
    {
      build: furnace,
      tier: 2
    }
  ]
});
var nuclearPower = new Technology({
  name: "Nuclear Power",
  price: { gems: 81500 },
  unlocksTechnologies: [],
  resourcesNeeded: [
    {
      item: concrete,
      amount: 1e3
    },
    {
      item: steel,
      amount: 1e3
    },
    {
      item: enrichedUranium,
      amount: 100
    }
  ],
  unlocks: [nuclearPowerPlant, emptyFuelCell, nuclearFuelCell]
});
var coalPower = new Technology({
  name: "Coal Power",
  price: { gems: 600 },
  unlocksTechnologies: [],
  resourcesNeeded: [
    {
      item: coal,
      amount: 1e3
    }
  ],
  unlocks: [coalPowerPlant]
});
var workshopUpgrade3 = new Technology({
  name: "Workshop Upgrade",
  price: { gems: 134400 },
  unlocksTechnologies: [],
  resourcesNeeded: [
    {
      item: turbocharger,
      amount: 1e3
    }
  ],
  unlocks: [
    {
      build: workshop,
      tier: 4
    }
  ]
});
var workshopUpgrade2 = new Technology({
  name: "Workshop Upgrade",
  price: { gems: 17400 },
  unlocksTechnologies: [workshopUpgrade3],
  resourcesNeeded: [
    {
      item: ironGear,
      amount: 5e3
    },
    {
      item: logicCircuit,
      amount: 2e3
    }
  ],
  unlocks: [
    {
      build: workshop,
      tier: 3
    }
  ]
});
var workshopUpgrade1 = new Technology({
  name: "Workshop Upgrade",
  price: { gems: 1200 },
  unlocksTechnologies: [workshopUpgrade2],
  resourcesNeeded: [
    {
      item: copperWire,
      amount: 1e3
    }
  ],
  unlocks: [
    {
      build: workshop,
      tier: 2
    }
  ]
});
var undergroundBeltUpgrade2 = new Technology({
  name: "Underground Belt Upgrade",
  price: { gems: 116100 },
  unlocksTechnologies: [],
  resourcesNeeded: [
    {
      item: turbocharger,
      amount: 200
    },
    {
      item: concrete,
      amount: 4e3
    }
  ],
  unlocks: [
    {
      build: undergroundBelt,
      tier: 2
    }
  ]
});
var undergroundBeltUpgrade1 = new Technology({
  name: "Underground Belt Upgrade",
  price: { gems: 25700 },
  unlocksTechnologies: [undergroundBeltUpgrade2],
  resourcesNeeded: [
    {
      item: nanoWire,
      amount: 100
    },
    {
      item: concrete,
      amount: 1e3
    }
  ],
  unlocks: [
    {
      build: undergroundBelt,
      tier: 1
    }
  ]
});
var speedUpgradeXII = new Technology({
  name: "Speed Upgrade XII",
  price: { gems: 17e5 },
  unlocksTechnologies: [],
  resourcesNeeded: [
    {
      item: ironGear,
      amount: 15e4
    },
    {
      item: earthToken,
      amount: 100
    }
  ],
  unlocks: [
    {
      build: conveyorBelt,
      tier: 12
    }
  ]
});
var speedUpgradeXI = new Technology({
  name: "Speed Upgrade XI",
  price: { gems: 898500 },
  unlocksTechnologies: [speedUpgradeXII],
  resourcesNeeded: [
    {
      item: ironGear,
      amount: 1e5
    },
    {
      item: earthToken,
      amount: 50
    }
  ],
  unlocks: [
    {
      build: conveyorBelt,
      tier: 11
    }
  ]
});
var speedUpgradeX = new Technology({
  name: "Speed Upgrade X",
  price: { gems: 94700 },
  unlocksTechnologies: [speedUpgradeXI],
  resourcesNeeded: [
    {
      item: ironGear,
      amount: 8e4
    }
  ],
  unlocks: [
    {
      build: conveyorBelt,
      tier: 10
    }
  ]
});
var speedUpgradeIX = new Technology({
  name: "Speed Upgrade IX",
  price: { gems: 71100 },
  unlocksTechnologies: [speedUpgradeX],
  resourcesNeeded: [
    {
      item: ironGear,
      amount: 6e4
    }
  ],
  unlocks: [
    {
      build: conveyorBelt,
      tier: 9
    }
  ]
});
var speedUpgradeVIII = new Technology({
  name: "Speed Upgrade VIII",
  price: { gems: 47400 },
  unlocksTechnologies: [speedUpgradeIX],
  resourcesNeeded: [
    {
      item: ironGear,
      amount: 4e4
    }
  ],
  unlocks: [
    {
      build: conveyorBelt,
      tier: 8
    }
  ]
});
var speedUpgradeVII = new Technology({
  name: "Speed Upgrade VII",
  price: { gems: 35600 },
  unlocksTechnologies: [speedUpgradeVIII],
  resourcesNeeded: [
    {
      item: ironGear,
      amount: 3e4
    }
  ],
  unlocks: [
    {
      build: conveyorBelt,
      tier: 7
    }
  ]
});
var speedUpgradeVI = new Technology({
  name: "Speed Upgrade VI",
  price: { gems: 23700 },
  unlocksTechnologies: [speedUpgradeVII],
  resourcesNeeded: [
    {
      item: ironGear,
      amount: 2e4
    }
  ],
  unlocks: [
    {
      build: conveyorBelt,
      tier: 6
    }
  ]
});
var speedUpgradeV = new Technology({
  name: "Speed Upgrade V",
  price: { gems: 11900 },
  unlocksTechnologies: [speedUpgradeVI],
  resourcesNeeded: [
    {
      item: ironGear,
      amount: 1e4
    }
  ],
  unlocks: [
    {
      build: conveyorBelt,
      tier: 5
    }
  ]
});
var speedUpgradeIV = new Technology({
  name: "Speed Upgrade IV",
  price: { gems: 6e3 },
  unlocksTechnologies: [speedUpgradeV],
  resourcesNeeded: [
    {
      item: ironGear,
      amount: 5e3
    }
  ],
  unlocks: [
    {
      build: conveyorBelt,
      tier: 4
    }
  ]
});
var speedUpgradeIII = new Technology({
  name: "Speed Upgrade III",
  price: { gems: 2400 },
  unlocksTechnologies: [speedUpgradeIV],
  resourcesNeeded: [
    {
      item: ironGear,
      amount: 2e3
    }
  ],
  unlocks: [
    {
      build: conveyorBelt,
      tier: 3
    }
  ]
});
var speedUpgradeII = new Technology({
  name: "Speed Upgrade II",
  price: { gems: 1200 },
  unlocksTechnologies: [speedUpgradeIII],
  resourcesNeeded: [
    {
      item: ironGear,
      amount: 1e3
    }
  ],
  unlocks: [
    {
      build: conveyorBelt,
      tier: 2
    }
  ]
});
var speedUpgradeI = new Technology({
  name: "Speed Upgrade I",
  price: { gems: 600 },
  unlocksTechnologies: [speedUpgradeII],
  resourcesNeeded: [
    {
      item: ironGear,
      amount: 500
    }
  ],
  unlocks: [
    {
      build: conveyorBelt,
      tier: 1
    }
  ]
});
var advancedBelts = new Technology({
  name: "Advanced Belts",
  price: { gems: 60 },
  unlocksTechnologies: [speedUpgradeI, undergroundBeltUpgrade1],
  resourcesNeeded: [
    {
      item: ironGear,
      amount: 50
    }
  ],
  unlocks: [
    {
      build: beltSplitter,
      tier: 0
    },
    {
      build: undergroundBelt,
      tier: 0
    }
  ]
});
var uraniumProcessing = new Technology({
  name: "Uranium Processing",
  price: { gems: 9100 },
  unlocksTechnologies: [nuclearPower],
  resourcesNeeded: [
    {
      item: tungstenCarbide,
      amount: 1e3
    }
  ],
  unlocks: [
    {
      build: uraniumExtractor,
      tier: 1
    },
    enrichedUranium
  ]
});
var extractorUpgrade4 = new Technology({
  name: "Extractor Upgrade",
  price: { gems: 312100 },
  unlocksTechnologies: [],
  resourcesNeeded: [
    {
      item: earthToken,
      amount: 20
    }
  ],
  unlocks: [
    {
      build: extractor,
      tier: 5
    }
  ]
});
var extractorUpgrade3 = new Technology({
  name: "Extractor Upgrade",
  price: { gems: 67200 },
  unlocksTechnologies: [extractorUpgrade4],
  resourcesNeeded: [
    {
      item: electricMotor,
      amount: 500
    }
  ],
  unlocks: [
    {
      build: extractor,
      tier: 4
    }
  ]
});
var extractorUpgrade2 = new Technology({
  name: "Extractor Upgrade",
  price: { gems: 6900 },
  unlocksTechnologies: [extractorUpgrade3],
  resourcesNeeded: [
    {
      item: metalFrame,
      amount: 1e3
    }
  ],
  unlocks: [
    {
      build: extractor,
      tier: 3
    }
  ]
});
var extractorUpgrade1 = new Technology({
  name: "Extractor Upgrade",
  price: { gems: 1200 },
  unlocksTechnologies: [extractorUpgrade2],
  resourcesNeeded: [
    {
      item: woodFrame,
      amount: 500
    }
  ],
  unlocks: [
    {
      build: extractor,
      tier: 2
    }
  ]
});
var earthTeleporter2 = new Technology({
  name: "Earth Teleporter",
  price: { gems: 156100 },
  unlocksTechnologies: [],
  resourcesNeeded: [
    {
      item: matterDuplicator,
      amount: 10
    }
  ],
  unlocks: [
    {
      build: earthTeleporter,
      tier: 1
    },
    earthToken
  ]
});
var matterDuplication = new Technology({
  name: "Matter Duplication",
  price: { gems: 145600 },
  unlocksTechnologies: [earthTeleporter2],
  resourcesNeeded: [
    {
      item: particleGlue,
      amount: 1e3
    },
    {
      item: quantumEntangler,
      amount: 50
    }
  ],
  unlocks: [matterDuplicator]
});
var quantumEntanglement = new Technology({
  name: "Quantum Entanglement",
  price: { gems: 51400 },
  unlocksTechnologies: [matterDuplication],
  resourcesNeeded: [
    {
      item: atomicLocator,
      amount: 40
    }
  ],
  unlocks: [magneticFieldGenerator, quantumEntangler, particleGlue]
});
var atomicLocator2 = new Technology({
  name: "Atomic Locator",
  price: { gems: 26500 },
  unlocksTechnologies: [quantumEntanglement],
  resourcesNeeded: [
    {
      item: superComputer,
      amount: 100
    }
  ],
  unlocks: [atomicLocator]
});
var energyCube2 = new Technology({
  name: "Energy Cube",
  price: { gems: 37600 },
  unlocksTechnologies: [],
  resourcesNeeded: [
    {
      item: battery,
      amount: 200
    },
    {
      item: industrialFrame,
      amount: 100
    }
  ],
  unlocks: [energyCube]
});
var matterCompressor2 = new Technology({
  name: "Matter Compressor",
  price: { gems: 48200 },
  unlocksTechnologies: [],
  resourcesNeeded: [
    {
      item: turbocharger,
      amount: 200
    },
    {
      item: industrialFrame,
      amount: 100
    }
  ],
  unlocks: [tank, matterCompressor]
});
var industrialFrame2 = new Technology({
  name: "Industrial Frame",
  price: { gems: 38300 },
  unlocksTechnologies: [matterCompressor2, energyCube2],
  resourcesNeeded: [
    {
      item: concrete,
      amount: 1e3
    },
    {
      item: metalFrame,
      amount: 1e3
    },
    {
      item: tungstenCarbide,
      amount: 1e3
    }
  ],
  unlocks: [industrialFrame]
});
var stabilizer2 = new Technology({
  name: "Stabilizer",
  price: { gems: 22300 },
  unlocksTechnologies: [industrialFrame2],
  resourcesNeeded: [
    {
      item: rotor,
      amount: 1e3
    }
  ],
  unlocks: [gyroscope, stabilizer]
});
var superComputers = new Technology({
  name: "Super Computers",
  price: { gems: 16800 },
  unlocksTechnologies: [stabilizer2, atomicLocator2],
  resourcesNeeded: [
    {
      item: computer,
      amount: 100
    },
    {
      item: turbocharger,
      amount: 100
    }
  ],
  unlocks: [superComputer]
});
var turbocharger2 = new Technology({
  name: "Turbocharger",
  price: { gems: 8500 },
  unlocksTechnologies: [superComputers],
  resourcesNeeded: [
    {
      item: tungstenCarbide,
      amount: 200
    },
    {
      item: nanoWire,
      amount: 200
    }
  ],
  unlocks: [coupler, turbocharger]
});
var tungstenProduction = new Technology({
  name: "Tungsten Production",
  price: { gems: 600 },
  unlocksTechnologies: [uraniumProcessing, turbocharger2, manufacturerUpgrade1],
  resourcesNeeded: [
    {
      item: wolframite,
      amount: 1e3
    }
  ],
  unlocks: [tungstenOre, tungstenCarbide]
});
var microscopes = new Technology({
  name: "Microscopes",
  price: { gems: 13300 },
  unlocksTechnologies: [tungstenProduction],
  resourcesNeeded: [
    {
      item: nanoWire,
      amount: 400
    }
  ],
  unlocks: [
    {
      build: manufacturer,
      tier: 1
    },
    condenserLens,
    electronMicroscope
  ]
});
var nanoWire2 = new Technology({
  name: "Nano Wire",
  price: { gems: 13500 },
  unlocksTechnologies: [microscopes],
  resourcesNeeded: [
    {
      item: electricMotor,
      amount: 100
    }
  ],
  unlocks: [glass, carbonFiber, nanoWire]
});
var motors = new Technology({
  name: "Motors",
  price: { gems: 8e3 },
  unlocksTechnologies: [nanoWire2],
  resourcesNeeded: [
    {
      item: battery,
      amount: 50
    },
    {
      item: steelRod,
      amount: 200
    }
  ],
  unlocks: [rotor, electricMotor]
});
var batteries = new Technology({
  name: "Batteries",
  price: { gems: 4600 },
  unlocksTechnologies: [motors],
  resourcesNeeded: [
    {
      item: graphite,
      amount: 400
    },
    {
      item: electromagnet,
      amount: 400
    }
  ],
  unlocks: [battery]
});
var concrete2 = new Technology({
  name: "Concrete",
  price: { gems: 2e3 },
  unlocksTechnologies: [forgeUpgrade1],
  resourcesNeeded: [
    {
      item: steel,
      amount: 200
    },
    {
      item: sand,
      amount: 1e3
    }
  ],
  unlocks: [steelRod, concrete]
});
var steelProduction = new Technology({
  name: "Steel Production",
  price: { gems: 3400 },
  unlocksTechnologies: [concrete2, batteries, industrialFactoryUpgrade1],
  resourcesNeeded: [
    {
      item: computer,
      amount: 100
    }
  ],
  unlocks: [
    {
      build: forge,
      tier: 1
    },
    graphite,
    steel
  ]
});
var goldVault2 = new Technology({
  name: "Gold Vault",
  price: { gems: null },
  unlocksTechnologies: [],
  resourcesNeeded: [
    {
      // @ts-ignore This is the only exception
      item: gold,
      amount: 2e4
    }
  ],
  unlocks: [goldVault]
});
var computers = new Technology({
  name: "Computers",
  price: { gems: 2100 },
  unlocksTechnologies: [steelProduction, goldVault2],
  resourcesNeeded: [
    {
      item: ironPlating,
      amount: 500
    },
    {
      item: logicCircuit,
      amount: 250
    }
  ],
  unlocks: [
    {
      build: industrialFactory,
      tier: 1
    },
    heatSink,
    computer
  ]
});
var circuitry = new Technology({
  name: "Circuitry",
  price: { gems: 1300 },
  unlocksTechnologies: [robotics, computers],
  resourcesNeeded: [
    {
      item: copperWire,
      amount: 500
    },
    {
      item: metalFrame,
      amount: 100
    }
  ],
  unlocks: [silicon, logicCircuit]
});
var sand2 = new Technology({
  name: "Sand",
  price: { gems: 100 },
  unlocksTechnologies: [],
  resourcesNeeded: [
    {
      item: stone,
      amount: 100
    }
  ],
  unlocks: [sand]
});
var metalFrames = new Technology({
  name: "Metal Frames",
  price: { gems: 800 },
  unlocksTechnologies: [storage2, sand2, circuitry, machineShopUpgrade1],
  resourcesNeeded: [
    {
      item: electromagnet,
      amount: 100
    }
  ],
  unlocks: [ironPlating, metalFrame]
});
var magnets = new Technology({
  name: "Magnets",
  price: { gems: 120 },
  unlocksTechnologies: [metalFrames, coalPower, furnaceUpgrade1],
  resourcesNeeded: [
    {
      item: copperWire,
      amount: 100
    }
  ],
  unlocks: [
    {
      build: machineShop,
      tier: 1
    },
    electromagnet
  ]
});
var metalWorking = new Technology({
  name: "Metal Working",
  price: { gems: 60 },
  unlocksTechnologies: [advancedBelts, magnets, workshopUpgrade1],
  resourcesNeeded: [
    {
      item: ironIngot,
      amount: 50
    },
    {
      item: copperIngot,
      amount: 50
    }
  ],
  unlocks: [copperWire, ironGear]
});
var smelting = new Technology({
  name: "Smelting",
  price: { gems: 50 },
  unlocksTechnologies: [extractorUpgrade1, metalWorking, decorativeTrees],
  resourcesNeeded: [
    {
      item: woodFrame,
      amount: 20
    }
  ],
  unlocks: [
    {
      build: furnace,
      tier: 1
    },
    ironIngot,
    copperIngot
  ]
});
var woodFrame2 = new Technology({
  name: "Wood Frame",
  price: { gems: 20 },
  unlocksTechnologies: [smelting],
  resourcesNeeded: [
    {
      item: woodPlank,
      amount: 20
    }
  ],
  unlocks: [woodFrame]
});
var woodPlanks = new Technology({
  name: "Wood Planks",
  price: { gems: 10 },
  unlocksTechnologies: [woodFrame2],
  resourcesNeeded: [
    {
      item: woodLog,
      amount: 10
    }
  ],
  unlocks: [
    {
      build: workshop,
      tier: 1
    },
    woodPlank
  ]
});
var startingTech = new Technology({
  name: "Starting Tech",
  price: { gems: null },
  unlocksTechnologies: [woodPlanks],
  resourcesNeeded: [],
  unlocks: [
    researchLab,
    {
      build: conveyorBelt,
      tier: 0
    },
    {
      build: extractor,
      tier: 1
    }
  ],
  technologyNeeded: null
});
woodPlanks.technologyNeeded = startingTech;
woodFrame2.technologyNeeded = woodPlanks;
smelting.technologyNeeded = woodFrame2;
metalWorking.technologyNeeded = smelting;
magnets.technologyNeeded = metalWorking;
metalFrames.technologyNeeded = magnets;
sand2.technologyNeeded = metalFrames;
circuitry.technologyNeeded = metalFrames;
computers.technologyNeeded = circuitry;
steelProduction.technologyNeeded = computers;
batteries.technologyNeeded = steelProduction;
motors.technologyNeeded = batteries;
nanoWire2.technologyNeeded = motors;
microscopes.technologyNeeded = nanoWire2;
tungstenProduction.technologyNeeded = microscopes;
turbocharger2.technologyNeeded = tungstenProduction;
superComputers.technologyNeeded = turbocharger2;
stabilizer2.technologyNeeded = superComputers;
industrialFrame2.technologyNeeded = stabilizer2;
matterCompressor2.technologyNeeded = industrialFrame2;
energyCube2.technologyNeeded = industrialFrame2;
atomicLocator2.technologyNeeded = superComputers;
quantumEntanglement.technologyNeeded = atomicLocator2;
matterDuplication.technologyNeeded = quantumEntanglement;
earthTeleporter2.technologyNeeded = matterDuplication;
extractorUpgrade1.technologyNeeded = smelting;
extractorUpgrade2.technologyNeeded = extractorUpgrade1;
extractorUpgrade3.technologyNeeded = extractorUpgrade2;
extractorUpgrade4.technologyNeeded = extractorUpgrade3;
uraniumProcessing.technologyNeeded = tungstenProduction;
advancedBelts.technologyNeeded = metalWorking;
speedUpgradeI.technologyNeeded = advancedBelts;
speedUpgradeII.technologyNeeded = speedUpgradeI;
speedUpgradeIII.technologyNeeded = speedUpgradeII;
speedUpgradeIV.technologyNeeded = speedUpgradeIII;
speedUpgradeV.technologyNeeded = speedUpgradeIV;
speedUpgradeVI.technologyNeeded = speedUpgradeV;
speedUpgradeVII.technologyNeeded = speedUpgradeVI;
speedUpgradeVIII.technologyNeeded = speedUpgradeVII;
speedUpgradeIX.technologyNeeded = speedUpgradeVIII;
speedUpgradeX.technologyNeeded = speedUpgradeIX;
speedUpgradeXI.technologyNeeded = speedUpgradeX;
speedUpgradeXII.technologyNeeded = speedUpgradeXI;
undergroundBeltUpgrade1.technologyNeeded = advancedBelts;
undergroundBeltUpgrade2.technologyNeeded = undergroundBeltUpgrade1;
workshopUpgrade1.technologyNeeded = metalWorking;
workshopUpgrade2.technologyNeeded = workshopUpgrade1;
workshopUpgrade3.technologyNeeded = workshopUpgrade2;
coalPower.technologyNeeded = magnets;
nuclearPower.technologyNeeded = uraniumProcessing;
furnaceUpgrade1.technologyNeeded = magnets;
furnaceUpgrade2.technologyNeeded = furnaceUpgrade1;
furnaceUpgrade3.technologyNeeded = furnaceUpgrade2;
storage2.technologyNeeded = metalFrames;
storageUpgrade1.technologyNeeded = storage2;
storageUpgrade2.technologyNeeded = storageUpgrade1;
storageUpgrade3.technologyNeeded = storageUpgrade2;
machineShopUpgrade1.technologyNeeded = metalFrames;
machineShopUpgrade2.technologyNeeded = machineShopUpgrade1;
machineShopUpgrade3.technologyNeeded = machineShopUpgrade2;
robotics.technologyNeeded = circuitry;
roboticArmUpgrade1.technologyNeeded = robotics;
roboticArmUpgrade2.technologyNeeded = roboticArmUpgrade1;
roboticArmUpgrade3.technologyNeeded = roboticArmUpgrade2;
decorativeTrees.technologyNeeded = smelting;
festiveDecorations.technologyNeeded = decorativeTrees;
decorativeRocks.technologyNeeded = decorativeTrees;
factoryDecorations.technologyNeeded = decorativeRocks;
decorativeWalls.technologyNeeded = factoryDecorations;
gearStatue2.technologyNeeded = decorativeTrees;
circuitStatue2.technologyNeeded = gearStatue2;
steelStatue2.technologyNeeded = circuitStatue2;
particleStatue2.technologyNeeded = steelStatue2;
earthStatue2.technologyNeeded = particleStatue2;
gemTree2.technologyNeeded = robotics;
industrialFactoryUpgrade1.technologyNeeded = steelProduction;
industrialFactoryUpgrade2.technologyNeeded = industrialFactoryUpgrade1;
industrialFactoryUpgrade3.technologyNeeded = industrialFactoryUpgrade2;
forgeUpgrade1.technologyNeeded = concrete2;
forgeUpgrade2.technologyNeeded = forgeUpgrade1;
forgeUpgrade3.technologyNeeded = forgeUpgrade2;
manufacturerUpgrade1.technologyNeeded = tungstenProduction;
manufacturerUpgrade2.technologyNeeded = manufacturerUpgrade1;
manufacturerUpgrade3.technologyNeeded = manufacturerUpgrade2;
goldVault2.technologyNeeded = computers;

// src/index.ts
var Space = {
  yg: {
    Discord: {
      id: "710012954736590908",
      bot: false,
      system: false,
      flags: 4194368,
      username: "space.yg",
      global_name: "Space.yg",
      discriminator: "0",
      avatar: "c6e77d918ce84cc49664ca505fab3727",
      banner: null,
      accentColor: null,
      avatar_decoration: null
    },
    GitHub: "space-yg",
    Instagram: "space.yg",
    X: "space_yag"
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Area,
  Base,
  Decoration,
  DistanceTier,
  DistanceTiers,
  Factory,
  Item,
  ItemCollector,
  OutputTier,
  OutputTiers,
  PowerPlant,
  Price,
  Seed,
  Space,
  SpeedTier,
  SpeedTiers,
  Storage,
  StorageTier,
  StorageTiers,
  Technology,
  Tier,
  Tiers,
  TransportationDistance,
  TransportationSpeed,
  currencies,
  decorations,
  factories,
  itemCollectors,
  items,
  others,
  powerPlants,
  technologies,
  transportations
});
