// selfmap/dist/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});// src/index.ts
var SelfMap = class {
  ////////////////////
  //// Properties ////
  ////////////////////
  /** The values associated with the map */
  #values;
  /** The unique property of the map */
  #uniqueProperty;
  /////////////////
  //// Getters ////
  /////////////////
  /** The unique property of the map */
  get uniqueProperty() {
    return this.#uniqueProperty;
  }
  /** The size of the map */
  get size() {
    return this.#values.length;
  }
  //////////////////////
  //// Constructors ////
  //////////////////////
  constructor(objects, uniqueProperty) {
    this.#values = objects;
    this.#uniqueProperty = uniqueProperty;
  }
  /////////////////////////
  //// Private Methods ////
  /////////////////////////
  /**
   * Check if a value is of type key
   * @param value The value to check
   * @returns `true` if value is of type key; `false` otherwise
   */
  #isKey(value) {
    return typeof value === "string" || typeof value === "number";
  }
  has(keyOrValue) {
    if (this.#isKey(keyOrValue)) return this.#values.some((ele) => ele[this.#uniqueProperty] === keyOrValue);
    else return this.#values.some((ele) => ele === keyOrValue);
  }
  /**
   * Add a value to the map
   * @param value The value to add
   * @returns The map itself
   */
  add(value) {
    if (!this.has(value)) {
      if (this.has(value[this.#uniqueProperty])) this.#values[this.#values.findIndex((ele) => ele[this.#uniqueProperty] === value[this.#uniqueProperty])] = value;
      else this.#values.push(value);
    }
    return this;
  }
  /**
   * Get the value associated with a key
   * @param key The key to use to get the value associated with
   * @returns The value; `undefined` if not found
   */
  get(key) {
    for (const ele of this.#values) if (ele[this.#uniqueProperty] === key) return ele;
    return void 0;
  }
  /**
   * Loop over each element in the map
   * @param callbackfn The callback function to call on each element
   * @param thisArg The value of `this` in the callback function
   */
  forEach(callbackfn, thisArg) {
    this.#values.forEach((ele) => callbackfn(ele, ele[this.#uniqueProperty], this), thisArg);
  }
  /**
   * Get all the keys in the map
   */
  keys() {
    return Iterator.from(this.#values.map((ele) => ele[this.#uniqueProperty]));
  }
  /**
   * Get all values in the map
   */
  values() {
    return Iterator.from(this.#values);
  }
  /**
   * Clear the map
   */
  clear() {
    this.#values = [];
  }
  delete(keyOrValue) {
    const isKey = this.#isKey(keyOrValue);
    let index;
    if (isKey) index = this.#values.findIndex((ele) => ele[this.#uniqueProperty] === keyOrValue);
    else index = this.#values.findIndex((ele) => ele === keyOrValue);
    if (index === -1) return void 0;
    if (isKey) return this.#values.splice(index, 1)[0];
    else return this.#values.splice(index, 1)[0][this.#uniqueProperty];
  }
  /**
   * Get all entries of the map
   */
  entries() {
    return Iterator.from(this.#values.map((ele) => [ele[this.#uniqueProperty], ele]));
  }
  /**
   * Loop over all the values in the map
   */
  [Symbol.iterator]() {
    return this.entries();
  }
  /**
   * Prints out this map
   */
  print() {
    if (this.#values.length !== 0) {
      console.log(`SelfMap(${this.size}) {`);
      this.#values.forEach((ele) => console.log(" ", ele[this.#uniqueProperty], "=>", ele));
      console.log("}");
    } else console.log(`SelfMap(${this.size}) {}`);
  }
  /**
   * Alias for {@link SelfMap.print print}
   */
  log() {
    this.print();
  }
  /**
   * A string representation of the map
   */
  toString() {
    if (this.#values.length === 0) return `SelfMap(${this.size}) {}`;
    let str = `SelfMap(${this.size}) {
`;
    this.#values.forEach((ele) => str += `  ${ele[this.#uniqueProperty]} => ${ele}
`);
    str += "}";
    return str;
  }
  /**
   * A string representation of the map
   */
  [Symbol.toStringTag]() {
    return this.toString();
  }
};


exports.default = SelfMap;

module.exports = exports.default;
