// builderment-classes/dir/index.js
export const Space = {
    yg: {
        Discord: {
            id: '710012954736590908',
            bot: false,
            system: false,
            flags: 4194368,
            username: 'space.yg',
            global_name: 'Space.yg',
            discriminator: '0',
            avatar: 'c6e77d918ce84cc49664ca505fab3727',
            banner: null,
            accentColor: null,
            avatar_decoration: null,
        },
        GitHub: "space-yg",
        Instagram: "space.yg",
        X: "space_yag",
    },
};
export { Area, } from "./classes/Area.js";
export { Base, } from "./classes/Base.js";
export { Currency } from "./classes/Currency.js";
export { Item, } from "./classes/Item.js";
export { Price, } from "./classes/Price.js";
export { Seed, } from "./classes/Seed.js";
export { Technology, } from "./classes/Technology.js";
export { Decoration, } from "./classes/buildings/Decoration.js";
export { ItemCollector, } from "./classes/buildings/ItemCollector.js";
export { PowerPlant, } from "./classes/buildings/PowerPlant.js";
export { Tier, } from "./classes/buildings/tiers/Tier.js";
export { Tiers, } from "./classes/buildings/tiers/Tiers.js";
export { DistanceTier, } from "./classes/buildings/tiers/distance-tiers/DistanceTier.js";
export { DistanceTiers, } from "./classes/buildings/tiers/distance-tiers/DistanceTiers.js";
export { TransportationDistance, } from "./classes/buildings/tiers/distance-tiers/TransportationDistance.js";
export { OutputTier, } from "./classes/buildings/tiers/output-tiers/OutputTier.js";
export { OutputTiers, } from "./classes/buildings/tiers/output-tiers/OutputTiers.js";
export { Factory, } from "./classes/buildings/tiers/output-tiers/Factory.js";
export { SpeedTier, } from "./classes/buildings/tiers/speed-tier/SpeedTier.js";
export { SpeedTiers, } from "./classes/buildings/tiers/speed-tier/SpeedTiers.js";
export { TransportationSpeed, } from "./classes/buildings/tiers/speed-tier/TransportationSpeed.js";
export { StorageTier, } from "./classes/buildings/tiers/storage-tiers/StorageTier.js";
export { StorageTiers, } from "./classes/buildings/tiers/storage-tiers/StorageTiers.js";
export { Storage, } from "./classes/buildings/tiers/storage-tiers/Storage.js";
export { Design, } from "./classes/designs/Design.js";
export { Requirements, RoboticArmTiers, } from "./classes/designs/Requirements.js";
export { NToM, } from "./classes/designs/NToM.js";
export { RoboticArmTierDesign, } from "./classes/designs/RoboticArmTierDesign.js";
import * as currencies_1 from "./objects/currencies.js";
export { currencies_1 as currencies };
import * as items_1 from "./objects/items.js";
export { items_1 as items };
import * as technologies_1 from "./objects/technologies.js";
export { technologies_1 as technologies };
import * as decorations_1 from "./objects/buildings/decorations.js";
export { decorations_1 as decorations };
import * as factories_1 from "./objects/buildings/factories.js";
export { factories_1 as factories };
import * as itemCollectors_1 from "./objects/buildings/item-collectors.js";
export { itemCollectors_1 as itemCollectors };
import * as powerPlants_1 from "./objects/buildings/power-plants.js";
export { powerPlants_1 as powerPlants };
import * as others_1 from "./objects/buildings/storages.js";
export { others_1 as others };
import * as transportations_1 from "./objects/buildings/transportations.js";
export { transportations_1 as transportations };
import * as balancers_1 from "./objects/designs/balancers.js";
export { balancers_1 as balancers };
import * as factorySplitters_1 from "./objects/designs/factory-splitters.js";
export { factorySplitters_1 as factorySplitters };
import * as labBalancers_1 from "./objects/designs/lab-balancers.js";
export { labBalancers_1 as labBalancers };
import * as splitters_1 from "./objects/designs/splitters.js";
export { splitters_1 as splitters };
import * as valves_1 from "./objects/designs/valves.js";
export { valves_1 as valves };
